---
title: "NK cells immunoediting"
author: "Aleksandr Bykov + code from Peter Repiscak"
date: '`r format(Sys.time(), "%d %B %Y")`'
output:
 html_document:
  theme: readable
  highlight: tango
  code_folding: hide
  code_download: true
#  toc: true
#  toc_depth: 2
#  toc_float: true
always_allow_html: true
params:
  config_file: "nk_tum_immunoediting_config.yaml"
---

# Project description 

Uncovering immune evasion mechanisms of leukemic cells from natural killer cells. In this sub-project RNA sequencing of leukemic cells is used to discover transcriptional changes resulting in NK cell-resistance. There is another sub-project using ATAC-seq (transposase-accessible chromatin) to epigenetic changes resulting in NK cell-resistance - this is not part of this report.

## Experimental design

Taken from discussion summary ('21009_Porject discussion summary.docx'): 

*	4 different cell lines (p185#13G, p185#13H, p185#15M, p185#15O). ADD description about cell line differences.
*	All conditions in technical triplicates and 2 biological replicates (biological replicates can be added any time, if there is a need).
* Technical replicate means that we already start the co-culture in a separate well.

![Biological and technical replicates](NKproject_bio_technical_replicates.png)

## Timeline of the experiment

*	On time point Day 4 we FACS sort the original tumor cells (like Day 0) and the tumor cells which were co-cultured for 4 days with NK cells for the following analyses: ATAC seq, RNA seq and barcode analysis.
Additionally we co-culture a certain amount of the same tumor cells again with NK cells for the analysis of Day 10.
*	On time point Day 10 we FACS sort the tumor cells which were co-cultured for 10 days with NK cells for the same analyses: ATAC seq, RNA seq and barcode analysis.

![Experiment Timepoints](NKproject_experiment_timepoints.png)

# RNA-seq data analysis {.tabset .tabset-pills}

RNA-seq methods: QuantSeq 3' mRNA-Seq - 1x50bp HiSeq 3000/4000. 

Pre-processing: ADD methods+versions for preprocessing
![Pre-processing pipeline](NKproject_preprocessing_pipeline.png)

Downstream analysis: ADD methods+versions for downstream!


[MultiQC report](NKproject_overall_multiqc.html)
ADD full multiqc report description. Few observations:

* 3' bias (expected for Quant-seq)
* EXP9.8 has longer reads (?) 120bp as opposed to 51bp for every other experiment. This is also reflected in a higher alignment to coding region. Further, a clear separation of EXP9.8 from other experiments is observed on the PCA plot. However, also EXP9.7 is somehow separated (along the PC2) from the other experiments. The quantification and alignment were done with the same settings all the experiments


```{r analysis_parameters, include=FALSE}
# Specifying parameters here ----
# remotes::install_github("rstudio/renv@v0.16.0")
if(!("renv" %in% installed.packages())){install.packages("renv", version="0.16.0")}  # docker has 0.15.4; change in production image! 
if(!("import" %in% installed.packages())){renv::install("import")}
if(!("yaml" %in% installed.packages())){renv::install("yaml")}
if(!("knitr" %in% installed.packages())){renv::install("knitr")}

# setting up initial parameters
base_dir <- "/home/rstudio/workspace/"
data_dir <- file.path(base_dir, "datasets") 
dir.create(data_dir)
annotation_dir = paste0(data_dir, "/annotations/")
dir.create(annotation_dir)
results_dir <- file.path(base_dir, "results_dir") 
dir.create(results_dir)
setwd(base_dir)

# project setup
# loading config file ----
config_file <- file.path(base_dir, "nk_tum_immunoediting_config.yaml")  
config <- yaml::read_yaml(config_file)

#project setup
project_name <- config$project_name
nthreads <- config$nthreads # e.g. 4 for furrr multisession

# this make take a while...
# https://rstudio.github.io/renv/reference/config.html
# getOption(x, default = NULL)
# renv::settings$use.cache()
# getOption('renv.config.pak.enabled')
# if the project does not automatically activate run:
if(Sys.getenv("RENV_PATHS_CACHE") != "/renv_cache") {Sys.setenv(RENV_PATHS_CACHE = "/renv_cache")}
if(Sys.getenv("RENV_PATHS_LIBRARY") != "/home/rstudio/renv_library") {Sys.setenv(RENV_PATHS_LIBRARY = "/home/rstudio/renv_library")}
#if(Sys.getenv("RENV_CONFIG_PAK_ENABLED") != "TRUE") {Sys.setenv(RENV_CONFIG_PAK_ENABLED = TRUE)} # add pak, targets and benchmarkme to docker!!!
# setting root dir
# 0. renv::activate
renv::activate(project = base_dir)

#IF packages are not loaded properly, use: 
# renv::restore()

# Only for manual installation of the project and recovery from the lock file.
# renv::init(project = base_dir, bare=TRUE, bioconductor = "3.16")
# FOR the first run restore environment and packages from renv.lock 
# 1. restore original environment
# renv::restore(project = base_dir, lockfile = file.path(base_dir, paste0(project_name, "_renv.lock")), prompt = FALSE)
# file.copy(from = file.path(base_dir, paste0(project_name, "_renv.lock")), to = file.path(base_dir, "renv.lock")) # to use with renv::diagnostics()

# for some reason apeglm package cannot be spanshotted to the renv.lock file. so, it must be installed manually.
# BiocManager::install("apeglm")

```

```{r initial_setup, include=FALSE}
# move to an internal hidden chunk?

set.seed(42)

import::from(knitr, opts_chunk)

# report options ----
# knitr::opts_chunk
opts_chunk$set(echo = TRUE,
                      eval = TRUE,
                      cache.lazy = FALSE,
                      message=FALSE,
                      warning=FALSE,
                      dev = "png")

# cache.path = file.path(results_dir,"report","cache/")
# fig.path = file.path(results_dir,"report","files/")                     
                     
options(width=100)

```


```{r run in the terminal, eval = FALSE, include = FALSE}
# run in terminal to generate report
rmarkdown::render(output_file = file.path(base_dir, "nk_tum_immunoediting.html"), 
                  input = here::here("nk_tum_immunoediting.Rmd"))
```

```{r loading libraries, message=FALSE, eval=TRUE, include=TRUE}
# importing only key functions that are actually used - not to polute namespace!
import::from(readr, read_csv)
import::from(magrittr, "%>%")
import::from(dplyr, mutate, select, filter, rename, arrange, desc, group_by, summarise, ungroup)  # dplyr_mutate = mutate
import::from(purrr, map)
import::from(future, plan, multisession, sequential)
import::from(furrr, furrr_options, future_map2)
import::from(ggplot2, .all=TRUE) # importing all as there is too many
import::from(grid, gpar) # needed in complexheatmap
import::from(kableExtra, kable_styling, kbl)
import::from(.from = SummarizedExperiment, colData, assay) # used in every .Rmd
import::from(.from = tableone, CreateTableOne)

```

## Preparing datasets 

```{r, child=here::here("00_prepare_datasets.Rmd"), include=TRUE, eval = TRUE}
```

## Exploratory Data Analysis - metadata

```{r, child=here::here("01_eda_metadata.Rmd"), include=TRUE, eval = TRUE}
```

## Exploratory Data Analysis - expression

```{r, child=here::here("02_eda_expression.Rmd"), include=TRUE, eval = TRUE}
```

## Results - TpNK_tp2_vs_Tonly_tp1 

```{r, child=here::here("03_results_TpNK_tp2_vs_Tonly_tp1_Aleks_AB_CD_separated.Rmd"), include=TRUE, eval = TRUE}
```

## Results - EXP9.8 comparing WT vs KO, IFNG and PrfKO_NK timepoints

```{r, child=here::here("04_exp9.8_KO_WT_stim_comparison.Rmd"), include=TRUE, eval = TRUE}
```

# sessionInfo
```{r include=TRUE, eval=FALSE, message=FALSE, warning=FALSE}
sessionInfo()

# saving session info
sink(file.path(results_dir, paste0("sessionInfo_", project_name,".txt")))
sessionInfo()
sink()

renv::snapshot(lockfile = file.path(base_dir, paste0(project_name, "_renv.lock")))
renv::status(lockfile = file.path(base_dir, paste0(project_name, "_renv.lock")))

```
