```{r loading libraries - prepare_datasets, message=FALSE, eval=TRUE, include=TRUE}
# importing only key functions that are actually used - not to polute namespace!
import::from(.from = readr, read_csv, cols)
import::from(magrittr, "%>%")
import::from(dplyr, mutate, select, filter, rename, arrange, desc, group_by, summarise, ungroup)  # dplyr_mutate = mutate
import::from(purrr, map)
import::from(future, plan, multisession, sequential)
import::from(furrr, furrr_options, future_map2)
import::from(ggplot2, .all=TRUE) # importing all as there is too many
import::from(grid, gpar) # needed in complexheatmap
import::from(kableExtra, kable_styling, kbl)

import::from(.from = GenomicFeatures, makeTxDbFromGFF)
import::from(.from = AnnotationDbi, annot_db_keys = keys, annot_db_select = select)
import::from(.from = DESeq2, .all=TRUE)
import::from(.from = tximport, tximport)

import::from(.from = here::here("utils/filterDatasets.R"), "filterDatasets", .character_only=TRUE) # used for filtering
import::from(.from = here::here("utils/generateEnsemblAnnotation.R"), "generateEnsemblAnnotation", .character_only=TRUE) # used for filtering

```

```{r experiment information, message=FALSE, eval=TRUE, include=TRUE}
experiment_name="nk_tum_immunoedit_complete" # this can be a subset analysis - e.g. just batch1,...
experiment_design="1"  # used only to construct initial dds object
abs_filt_samples=3

# parameters for annotation
biomart_host = "http://nov2020.archive.ensembl.org"
#biomart_host = "https://www.ensembl.org"
biomart_Ens_version = "Ensembl Genes 102"
biomart_dataset="mmusculus_gene_ensembl"

# directories
output_dir = paste0("results_dir", "/",experiment_name,"/")
dir.create(output_dir)
```

Preparing metadata and expression data and combining into a dds (DESeqDataSet object) that is ready for analyses. dds object is further pre-filtered to remove lowly expressed genes (FPM (fragments per million mapped fragments) > 1 in at least `r abs_filt_samples` samples). Salmon quantification (quant.sf) transcript abundance estimates (transcript expression) is imported and converted to gene expression using tximport.

```{r preparing metadata and salmon files, eval=TRUE, include=TRUE, message=FALSE}
# non relevant anymore, need to re-write for GEO pulls

# # Preparing metadata
# rnaseq_metadata_file <- here::here(config$metadata$rnaseq$rnaseq_metadata)
# rnaseq_metadata_df_raw <- readr::read_csv(file = rnaseq_metadata_file)  # 133 samples; 42 samples from other project?
# quantseq_files <- here::here(config$metadata$rnaseq$rnaseq_salmon_files)
# 
# # salmon quant files
# quant_files_md5sums_raw <- readr::read_table(file = quantseq_files, col_names = c("md5sum", "file_path"))
# quant_files_md5sums <- quant_files_md5sums_raw %>%
#   dplyr::mutate(absolute_quant_files_path = gsub(pattern = "\\./", replacement = paste0(dirname(quantseq_files), "/"), file_path),
#                 sample_name = gsub(pattern = "(\\./)(S_.+)(_0_.+)", replacement = "\\2", file_path)) %>%
#   # for batch3
#   dplyr::mutate(sample_name = gsub(pattern = "(\\./)(R.+)(_S.+)", replacement = "\\2", sample_name)) %>%
#   # EK were samples from other experiment - they were removed from quant.sf list, but in case remove them here as well!
#   dplyr::filter(!grepl(pattern = "\\./EK.+", x = sample_name)) %>%
#   dplyr::mutate(bsf_sample_name = gsub(pattern = "(\\./)(S_.+)(_quant/quant.sf)", replacement = "\\2", x = file_path)) %>%
#   # for batch 3
#   dplyr::mutate(bsf_sample_name = gsub(pattern = "(\\./)(R.+_S.+)(_quant/quant.sf)", replacement = "\\2", x = bsf_sample_name))
# 
# # experiment info - match experiment to run
# experiment_info_batch1_2_raw <- readr::read_csv("datasets/metadata/rnaseq_merge_sample_sheet.csv", col_types =cols(.default = "c"))
# experiment_info_batch3_raw <- readr::read_csv("datasets/metadata/rnaseq_sample_sheet_batch3.csv", col_types = cols(.default = "c"))
# 
# experiment_info_raw <- dplyr::bind_rows(experiment_info_batch1_2_raw, experiment_info_batch3_raw)
# 
# # EK samples are from different experiment
# experiment_info <- experiment_info_raw %>%
#   dplyr::filter(!grepl(pattern = "EK.+", x = original_sample_name)) %>%
#   dplyr::rename(library_name = experiment,
#                 bsf_sample_name = sample_name) %>%
#   # fixing library_name for experiment EXP9.8
#   dplyr::mutate(library_name = dplyr::if_else(library_name == "BSF_1241", "R0128_L5464", library_name)) %>%
#   dplyr::mutate(sample_name_unique = dplyr::case_when(grepl(pattern = "^R", x=original_sample_name) ~ paste0(original_sample_name, "_", library_name),
#                                                       TRUE ~ paste0("S_", original_sample_name, "_", library_name)))
# # loading metadata and adding experiment and salmon data
# rnaseq_metadata_df <- rnaseq_metadata_df_raw %>%
#   dplyr::mutate(condition_complete = condition) %>%
#   dplyr::mutate(condition = gsub(pattern = "(.+)(_Day.+)", replacement = "\\1", condition_complete)) %>%
#   dplyr::mutate(condition = gsub(pattern = "\\+", replacement = "_plus_", condition),
#                 timepoint_cell_harvesting = gsub(pattern = " ", replacement = "_", timepoint_cell_harvesting),
#                 experiment = gsub(pattern = " ", replacement = "", original_experiment)) %>%
#   dplyr::mutate(sample_name_unique = dplyr::case_when(grepl(pattern = "^R", x=sample_name) ~ paste0(sample_name, "_", library_name),
#                    TRUE ~ paste0("S_", sample_name, "_", library_name)))
# 
# quant_files_md5sums_annot <- quant_files_md5sums %>%
#   dplyr::left_join(., experiment_info, by = "bsf_sample_name") %>%
#   dplyr::select(-sample_name, -library_name)
# 
# rnaseq_metadata_df_annot <- rnaseq_metadata_df %>%
#   dplyr::left_join(., quant_files_md5sums_annot, by="sample_name_unique") %>%
#   dplyr::mutate(original_sample_name = sample_name) %>%
#   dplyr::mutate(sample_name = sample_name_unique) %>%
#   dplyr::mutate(filenames = sample_name_unique)
# 
# #table(rnaseq_metadata_df_annot$original_experiment)
# #table(rnaseq_metadata_df_annot$cell_line_label)
# rnaseq_metadata_df_annot_reduced <- rnaseq_metadata_df_annot %>%
#   #dplyr::filter(original_experiment == "EXP9_6") %>%
#   #dplyr::filter(condition %in% c("Tumor_only", "Tumor_plus_NK_Day14")) %>%
#   dplyr::mutate(technical_replicate = factor(technical_replicate, levels = c("1", "2", "3")),
#                 condition = factor(condition, levels = c("Tumor_only", "Tumor_plus_NK", "Tumor_plus_IFNg", "Tumor_plus_WT_NK", "Tumor_plus_PrfKO_NK")),
#                 timepoint_cell_harvesting = factor(timepoint_cell_harvesting, levels = c("timepoint_0", "timepoint_1", "timepoint_2")),
#                 cell_line_label = as.factor(cell_line_label),
#                 experiment = factor(experiment, levels = c("EXP9.4", "EXP9.5", "EXP9.6", "EXP9.7", "EXP9.8")),
#                 condition_tp = paste0(condition,"_",timepoint_cell_harvesting)) %>%
#   dplyr::mutate(condition_tp = factor(condition_tp, levels = c("Tumor_only_timepoint_0", 
#                                                                "Tumor_only_timepoint_1", "Tumor_only_timepoint_2",
#                                                                "Tumor_plus_NK_timepoint_1", "Tumor_plus_NK_timepoint_2",
#                                                                "Tumor_plus_WT_NK_timepoint_2",
#                                                                "Tumor_plus_IFNg_timepoint_2", "Tumor_plus_PrfKO_NK_timepoint_2"))) %>%
#   #fixing day_cell_harvesting 
#   dplyr::mutate(day_cell_harvesting = gsub(pattern = " ", replacement = "", day_cell_harvesting)) %>%
#   dplyr::mutate(day_cell_harvesting = factor(day_cell_harvesting, levels=c("Day0", "Day4", "Day6", "Day10", "Day14", "Day16")))
# 
# metadata <- rnaseq_metadata_df_annot_reduced #cond_data
# #rm(reseq_metadata)
# rownames(metadata) <- metadata$filenames
# 
# # use alternative sample names for multiQC
# # renaming for the multiQC report
# # metadata_multiqc <- metadata %>%
# #   dplyr::select(bsf_sample_name, experiment, technical_replicate, cell_line_label, condition, timepoint_cell_harvesting) %>%
# #   dplyr::mutate(cell_line_label = paste0("cl", cell_line_label), 
# #                 condition = gsub(pattern = "Tumor", replacement = "T", x = condition),
# #                 timepoint_cell_harvesting = gsub(pattern = "timepoint", replacement = "tp", x=timepoint_cell_harvesting),
# #                 technical_replicate = paste0("repl",technical_replicate)) %>%
# #   dplyr::mutate(alt_name = paste(experiment, cell_line_label, technical_replicate, condition, timepoint_cell_harvesting, sep="-")) %>%
# #   dplyr::select(bsf_sample_name, alt_name)
# # 
# # readr::write_tsv(metadata_multiqc, file = paste0(OUTPUT_DIR, experiment_name, "_metadata_multiqc.tsv"))

```

```{r creating tx2gene, eval=TRUE, include=TRUE, message=FALSE}
# DELETE the creation of the tx2gene - use only the final tx2gene table


# gtf_file <- file.path(annotation_dir, "Mus_musculus.GRCm38.102.gtf") # needed only once to generate tx2gene.tsv
 tx2gene_file <- file.path(annotation_dir, "EnsDb.Mmusculus.v102_tx2gene.tsv")
# tx2gene_file_md5sum <- file.path(annotation_dir, "EnsDb.Mmusculus.v102_tx2gene.md5sum")
# 
# if (!file.exists(tx2gene_file)) {
#   message(tx2gene_file, " does not exist: generating!")
#   txDB <- GenomicFeatures::makeTxDbFromGFF(gtf_file, format="gtf")
#   
#   # AnnotationDbi::keys
#   tx_name <- annot_db_keys(txDB, keytype="TXNAME")
#   
#   # AnnotationDbi::select
#   tx2gene_gtf <- annot_db_select(txDB,
#                                  keys = tx_name,
#                                  columns="GENEID",
#                                  keytype = "TXNAME")
#   
#   
#   # saving tx2gene into destdir
#   write.table(tx2gene_gtf,
#               file = tx2gene_file,
#               sep = "\t",
#               col.names = FALSE,
#               row.names = FALSE,
#               quote = FALSE)
#   
#   tx2gene_md5sum <- digest::digest(tx2gene_file, file=TRUE, algo="md5")
#   write.table(paste(tx2gene_md5sum, basename(tx2gene_file), collapse = "\t"),
#               file = tx2gene_file_md5sum,
#               sep = "\t",
#               col.names = FALSE,
#               row.names = FALSE,
#               quote = FALSE)
#   
# } else {
#   message(tx2gene_file, " exists: loading!")
  tx2gene <- readr::read_tsv(tx2gene_file,
                             col_names = c("tx_id", "gene_id"),
                             show_col_types = FALSE)
# }

```

```{r preparing datasets, eval=TRUE, include=TRUE, message=FALSE}
precomputed_objects_filename <- paste0(experiment_name, "_dds_objects.RData")
precomputed_objects_file <- file.path(output_dir, precomputed_objects_filename)

precomputed_transf_objects_filename <- paste0(experiment_name, "_log2_vsd_filt_objects.RData")
precomputed_transf_objects_file <- file.path(output_dir, precomputed_transf_objects_filename)

if(!file.exists(precomputed_objects_file)){
  cat("RNA objects file '", precomputed_objects_filename, "' does not exist in the OUTPUT_DIR...\n")
  cat("generating", precomputed_objects_filename, "\n")
  # Generating dds and filtered dds objects ----
  #require(DESeq2)
  # here there is a single sample so we use ~1.
  # expect a warning that there is only a single sample...
  
  # generating dds object ----
  # txi_object <- generateTxi(metadata_object=metadata,
  #                           column_names=c("absolute_quant_files_path", "filenames"),
  #                           tx2gene = tx2gene)
  
  quant_files <- metadata$absolute_quant_files_path
  names(quant_files) <- metadata$sample_name
  
  txi_object <- tximport::tximport(quant_files,
                         type="salmon",
                         tx2gene = tx2gene,
                         ignoreTxVersion = TRUE)
  
  sum(rownames(metadata) == colnames(txi_object$counts)) # check if names match between metadata and counts data
  cat("... txi_object\n")
  
  dds <- DESeq2::DESeqDataSetFromTximport(txi_object, 
                                          colData = metadata, 
                                          design = as.formula(paste0("~", experiment_design)))
  
  cat("... dds\n")
  dds <- DESeq2::estimateSizeFactors(dds)
  dds <- DESeq2::DESeq(dds, minReplicatesForReplace=Inf) # do not replace outliers based on replicates
  
  # filtering dds
  cat("... dds filtering\n")
  dds_filt <- filterDatasets(dds, abs_filt = TRUE, 
                             abs_filt_samples = abs_filt_samples) # at least in N samples, which is a smallest group size
  dds_filt <- DESeq2::estimateSizeFactors(dds_filt)

  # Generate annotation file using biomaRt
  #FIXME:
  # Issues with Curl when using https:// and http:// throws a warning:
  #  Warning: Ensembl will soon enforce the use of https.
  #   Ensure the 'host' argument includes "https://"
  cat("... fetching annotation from biomart\n")
  ensemblAnnot <- generateEnsemblAnnotation(ensembl_ids = rownames(dds),
                                            host=biomart_host,
                                            version=biomart_Ens_version,
                                            dataset=biomart_dataset)
  
  cat("saving...", precomputed_objects_file, "\n")
  cat("...into:", output_dir, "\n")
  save(dds, dds_filt, ensemblAnnot,  file = precomputed_objects_file)
  
} else{
  cat("RNA objects file '", precomputed_objects_filename, "' exist...loading\n")
  load(precomputed_objects_file)
}

#transformation after filtering
if(!file.exists(precomputed_transf_objects_file)){
  # transformation
  log2_norm_filt <- DESeq2::normTransform(dds_filt)
  vsd_filt <- DESeq2::vst(dds_filt, blind = TRUE) # blind = TRUE for QC
  #rld_filt <- rlog(dds_filt, blind = TRUE)
  #rld_filt - too big for >100 samples; skipping!
  #rld_filt <- rlog(dds_filt, blind = FALSE) # not blind to batch effects
  save(log2_norm_filt, vsd_filt,   
       file = precomputed_transf_objects_file)
} else{
  load(precomputed_transf_objects_file)
}

```

Parameters for dataset preparation:

| Parameter           | Value                   | Decription                                               |
|------------------|---------------------|--------------------------------------|
| experiment_name     | `r experiment_name`     | this can be a subset analysis - e.g. just batch1,...     |
| experiment_design   | `r experiment_design`   | used only to construct initial dds object                |
| abs_filt_samples    | `r abs_filt_samples`    | \# at least in N samples, which is a smallest group size |
| biomart_host        | `r biomart_host`        |                                                          |
| biomart_Ens_version | `r biomart_Ens_version` |                                                          |
| biomart_dataset     | `r biomart_dataset`     |                                                          |
|                     |                         |                                                          |


Objects overview:
```{r dataset objects overview, eval=TRUE, include=TRUE, message=FALSE}
dds_objects_overview <- data.frame(dds_object = c("dds", "dds_filt"),
                                   n_samples = c(ncol(dds), ncol(dds_filt)),
                                   n_genes = c(nrow(dds), nrow(dds_filt)))

dds_objects_overview %>%
  kableExtra::kbl(caption = "dds objects overview") %>% 
  kableExtra::kable_classic(full_width = T)

# DT::datatable(dds_objects_overview,
#           options = list(pageLength = 10),
#           caption = 'Table 1: Samples in validation set')
```
