# Copied completely from the "Prepare dataset.Rmd"
# Trying to keep everythin in one place
# The analysis is similar to the previous version, however here I don't do the correction for the batch correction between AB and CD
```{r 03_loading libraries - prepare_datasets, message=FALSE, eval=TRUE, include=TRUE}
# importing only key functions that are actually used - not to polute namespace!
import::from(.from = readr, read_csv, cols)
import::from(magrittr, "%>%")
import::from(dplyr, mutate, select, filter, rename, arrange, desc, group_by, summarise, ungroup)  # dplyr_mutate = mutate
import::from(purrr, map)
import::from(future, plan, multisession, sequential)
import::from(furrr, furrr_options, future_map2)
import::from(ggplot2, .all=TRUE) # importing all as there is too many
import::from(grid, gpar) # needed in complexheatmap
import::from(kableExtra, kable_styling, kbl)

import::from(.from = GenomicFeatures, makeTxDbFromGFF)
import::from(.from = AnnotationDbi, annot_db_keys = keys, annot_db_select = select)
import::from(.from = DESeq2, .all=TRUE)
import::from(.from = tximport, tximport)

import::from(.from = here::here("utils/filterDatasets.R"), "filterDatasets", .character_only=TRUE) # used for filtering
import::from(.from = here::here("utils/generateEnsemblAnnotation.R"), "generateEnsemblAnnotation", .character_only=TRUE) # used for filtering
import::from(.from = here::here("utils/generateResults.R"), "meanExprsPerGroup", "generateResults_upd", .character_only=TRUE) # used for filtering

```

```{r 03 experiment information, message=FALSE, eval=TRUE, include=TRUE}
#project setup
# loading config file ----
config_file <- file.path( "nk_tum_immunoediting_config.yaml")  # params$config_file; params not found? ; change to project_config.yaml
config <- yaml::read_yaml(config_file)
#config <- yaml::read_yaml(params$config_file)

experiment_name="nk_tum_immunoedit_complete" # this can be a subset analysis - e.g. just batch1,...
#experiment_name="nk_tum_immunoedit_complete_ab2_AB_CD_sep" # this can be a subset analysis - e.g. just batch1,...
experiment_design="1"  # used only to construct initial dds object
abs_filt_samples=3

# parameters for annotation
biomart_host = "http://nov2020.archive.ensembl.org"
#biomart_host = "https://www.ensembl.org"
biomart_Ens_version = "Ensembl Genes 102"
biomart_dataset="mmusculus_gene_ensembl"

# directories
output_dir = paste0("./results_dir", "/",experiment_name,"/")
dir.create(output_dir)
```

Preparing metadata and expression data and combining into a dds (DESeqDataSet object) that is ready for analyses. dds object is further pre-filtered to remove lowly expressed genes (FPM (fragments per million mapped fragments) > 1 in at least `r abs_filt_samples` samples). Salmon quantification (quant.sf) transcript abundance estimates (transcript expression) is imported and converted to gene expression using tximport.

```{r 03 preparing metadata and salmon files, eval=TRUE, include=TRUE, message=FALSE}
# Preparing metadata
rnaseq_metadata_file <- here::here(config$metadata$rnaseq$rnaseq_metadata)
rnaseq_metadata_df_raw <- readr::read_csv(file = rnaseq_metadata_file)  # 133 samples; 42 samples from other project?
quantseq_files <- here::here(config$metadata$rnaseq$rnaseq_salmon_files)

# salmon quant files
quant_files_md5sums_raw <- readr::read_table(file = quantseq_files, col_names = c("md5sum", "file_path"))
quant_files_md5sums <- quant_files_md5sums_raw %>%
  dplyr::mutate(absolute_quant_files_path = gsub(pattern = "\\./", replacement = paste0(dirname(quantseq_files), "/"), file_path),
                sample_name = gsub(pattern = "(\\./)(S_.+)(_0_.+)", replacement = "\\2", file_path)) %>%
  # for batch3
  dplyr::mutate(sample_name = gsub(pattern = "(\\./)(R.+)(_S.+)", replacement = "\\2", sample_name)) %>%
  # EK were samples from other experiment - they were removed from quant.sf list, but in case remove them here as well!
  dplyr::filter(!grepl(pattern = "\\./EK.+", x = sample_name)) %>%
  dplyr::mutate(bsf_sample_name = gsub(pattern = "(\\./)(S_.+)(_quant/quant.sf)", replacement = "\\2", x = file_path)) %>%
  # for batch 3
  dplyr::mutate(bsf_sample_name = gsub(pattern = "(\\./)(R.+_S.+)(_quant/quant.sf)", replacement = "\\2", x = bsf_sample_name))

# experiment info - match experiment to run
experiment_info_batch1_2_raw <- readr::read_csv("datasets/metadata/rnaseq_merge_sample_sheet.csv", col_types =cols(.default = "c"))
experiment_info_batch3_raw <- readr::read_csv("datasets/metadata/rnaseq_sample_sheet_batch3.csv", col_types = cols(.default = "c"))

experiment_info_raw <- dplyr::bind_rows(experiment_info_batch1_2_raw, experiment_info_batch3_raw)

# EK samples are from different experiment
experiment_info <- experiment_info_raw %>%
  dplyr::filter(!grepl(pattern = "EK.+", x = original_sample_name)) %>%
  dplyr::rename(library_name = experiment,
                bsf_sample_name = sample_name) %>%
  # fixing library_name for experiment EXP9.8
  dplyr::mutate(library_name = dplyr::if_else(library_name == "BSF_1241", "R0128_L5464", library_name)) %>%
  dplyr::mutate(sample_name_unique = dplyr::case_when(grepl(pattern = "^R", x=original_sample_name) ~ paste0(original_sample_name, "_", library_name),
                                                      TRUE ~ paste0("S_", original_sample_name, "_", library_name)))
# loading metadata and adding experiment and salmon data
rnaseq_metadata_df <- rnaseq_metadata_df_raw %>%
  dplyr::mutate(condition_complete = condition) %>%
  dplyr::mutate(condition = gsub(pattern = "(.+)(_Day.+)", replacement = "\\1", condition_complete)) %>%
  dplyr::mutate(condition = gsub(pattern = "\\+", replacement = "_plus_", condition),
                timepoint_cell_harvesting = gsub(pattern = " ", replacement = "_", timepoint_cell_harvesting),
                experiment = gsub(pattern = " ", replacement = "", original_experiment)) %>%
  dplyr::mutate(sample_name_unique = dplyr::case_when(grepl(pattern = "^R", x=sample_name) ~ paste0(sample_name, "_", library_name),
                   TRUE ~ paste0("S_", sample_name, "_", library_name)))

quant_files_md5sums_annot <- quant_files_md5sums %>%
  dplyr::left_join(., experiment_info, by = "bsf_sample_name") %>%
  dplyr::select(-sample_name, -library_name)

rnaseq_metadata_df_annot <- rnaseq_metadata_df %>%
  dplyr::left_join(., quant_files_md5sums_annot, by="sample_name_unique") %>%
  dplyr::mutate(original_sample_name = sample_name) %>%
  dplyr::mutate(sample_name = sample_name_unique) %>%
  dplyr::mutate(filenames = sample_name_unique)

#table(rnaseq_metadata_df_annot$original_experiment)
#table(rnaseq_metadata_df_annot$cell_line_label)
rnaseq_metadata_df_annot_reduced <- rnaseq_metadata_df_annot %>%
  #dplyr::filter(original_experiment == "EXP9_6") %>%
  #dplyr::filter(condition %in% c("Tumor_only", "Tumor_plus_NK_Day14")) %>%
  dplyr::mutate(technical_replicate = factor(technical_replicate, levels = c("1", "2", "3")),
                condition = factor(condition, levels = c("Tumor_only", "Tumor_plus_NK", "Tumor_plus_IFNg", "Tumor_plus_WT_NK", "Tumor_plus_PrfKO_NK")),
                timepoint_cell_harvesting = factor(timepoint_cell_harvesting, levels = c("timepoint_0", "timepoint_1", "timepoint_2")),
                cell_line_label = as.factor(cell_line_label),
                experiment = factor(experiment, levels = c("EXP9.4", "EXP9.5", "EXP9.6", "EXP9.7", "EXP9.8")),
                condition_tp = paste0(condition,"_",timepoint_cell_harvesting)) %>%
  dplyr::mutate(condition_tp = factor(condition_tp, levels = c("Tumor_only_timepoint_0", 
                                                               "Tumor_only_timepoint_1", "Tumor_only_timepoint_2",
                                                               "Tumor_plus_NK_timepoint_1", "Tumor_plus_NK_timepoint_2",
                                                               "Tumor_plus_WT_NK_timepoint_2",
                                                               "Tumor_plus_IFNg_timepoint_2", "Tumor_plus_PrfKO_NK_timepoint_2"))) %>%
  #fixing day_cell_harvesting 
  dplyr::mutate(day_cell_harvesting = gsub(pattern = " ", replacement = "", day_cell_harvesting)) %>%
  dplyr::mutate(day_cell_harvesting = factor(day_cell_harvesting, levels=c("Day0", "Day4", "Day6", "Day10", "Day14", "Day16")))

metadata <- rnaseq_metadata_df_annot_reduced #cond_data
#rm(reseq_metadata)
rownames(metadata) <- metadata$filenames

# use alternative sample names for multiQC
# renaming for the multiQC report
# metadata_multiqc <- metadata %>%
#   dplyr::select(bsf_sample_name, experiment, technical_replicate, cell_line_label, condition, timepoint_cell_harvesting) %>%
#   dplyr::mutate(cell_line_label = paste0("cl", cell_line_label), 
#                 condition = gsub(pattern = "Tumor", replacement = "T", x = condition),
#                 timepoint_cell_harvesting = gsub(pattern = "timepoint", replacement = "tp", x=timepoint_cell_harvesting),
#                 technical_replicate = paste0("repl",technical_replicate)) %>%
#   dplyr::mutate(alt_name = paste(experiment, cell_line_label, technical_replicate, condition, timepoint_cell_harvesting, sep="-")) %>%
#   dplyr::select(bsf_sample_name, alt_name)
# 
# readr::write_tsv(metadata_multiqc, file = paste0(OUTPUT_DIR, experiment_name, "_metadata_multiqc.tsv"))

```

```{r 03 creating tx2gene, eval=TRUE, include=TRUE, message=FALSE}
#TODO:
# [ ] - add md5sums for tx2gene_file

data_dir <- file.path("datasets") 
annotation_dir = paste0(data_dir, "/annotations/")

gtf_file <- file.path(annotation_dir, "Mus_musculus.GRCm38.102.gtf") # needed only once to generate tx2gene.tsv
tx2gene_file <- file.path(annotation_dir, "EnsDb.Mmusculus.v102_tx2gene.tsv")
tx2gene_file_md5sum <- file.path(annotation_dir, "EnsDb.Mmusculus.v102_tx2gene.md5sum")

if (!file.exists(tx2gene_file)) {
  message(tx2gene_file, " does not exist: generating!")
  txDB <- GenomicFeatures::makeTxDbFromGFF(gtf_file, format="gtf")
  
  # AnnotationDbi::keys
  tx_name <- annot_db_keys(txDB, keytype="TXNAME")
  
  # AnnotationDbi::select
  tx2gene_gtf <- annot_db_select(txDB,
                                 keys = tx_name,
                                 columns="GENEID",
                                 keytype = "TXNAME")
  
  
  # saving tx2gene into destdir. Potentially there is a typo here. The name should be tx2gene, not _gtf
  write.table(tx2gene_gtf,
              file = tx2gene_file,
              sep = "\t",
              col.names = FALSE,
              row.names = FALSE,
              quote = FALSE)
  
  tx2gene_md5sum <- digest::digest(tx2gene_file, file=TRUE, algo="md5")
  write.table(paste(tx2gene_md5sum, basename(tx2gene_file), collapse = "\t"),
              file = tx2gene_file_md5sum,
              sep = "\t",
              col.names = FALSE,
              row.names = FALSE,
              quote = FALSE)
  
} else {
  message(tx2gene_file, " exists: loading!")
  tx2gene <- readr::read_tsv(tx2gene_file,
                             col_names = c("tx_id", "gene_id"),
                             show_col_types = FALSE)
}

```

```{r 03 preparing datasets, eval=TRUE, include=TRUE, message=FALSE}
precomputed_objects_filename <- paste0("nk_tum_immunoedit_complete", "_dds_objects.RData")
precomputed_objects_file <- file.path(output_dir, precomputed_objects_filename)

precomputed_transf_objects_filename <- paste0("nk_tum_immunoedit_complete", "_log2_vsd_filt_objects.RData")
precomputed_transf_objects_file <- file.path(output_dir, precomputed_transf_objects_filename)

if(!file.exists(precomputed_objects_file)){
  cat("RNA objects file '", precomputed_objects_filename, "' does not exist in the OUTPUT_DIR...\n")
  cat("generating", precomputed_objects_filename, "\n")
  # Generating dds and filtered dds objects ----
  #require(DESeq2)
  # here there is a single sample so we use ~1.
  # expect a warning that there is only a single sample...
  
  # generating dds object ----
  # txi_object <- generateTxi(metadata_object=metadata,
  #                           column_names=c("absolute_quant_files_path", "filenames"),
  #                           tx2gene = tx2gene)
  
  quant_files <- metadata$absolute_quant_files_path
  names(quant_files) <- metadata$sample_name
  
  txi_object <- tximport::tximport(quant_files,
                         type="salmon",
                         tx2gene = tx2gene,
                         ignoreTxVersion = TRUE)
  
  sum(rownames(metadata) == colnames(txi_object$counts)) # check if names match between metadata and counts data
  cat("... txi_object\n")
  
  dds <- DESeq2::DESeqDataSetFromTximport(txi_object, 
                                          colData = metadata, 
                                          design = as.formula(paste0("~", experiment_design)))
  
  cat("... dds\n")
  dds <- DESeq2::estimateSizeFactors(dds) #	Isn't the data already normalized by Salmon?
  dds <- DESeq2::DESeq(dds, minReplicatesForReplace=Inf) # do not replace outliers based on replicates
  
  # filtering dds
  cat("... dds filtering\n")
  dds_filt <- filterDatasets(dds, abs_filt = TRUE, 
                             abs_filt_samples = abs_filt_samples) # at least in N samples, which is a smallest group size
  dds_filt <- DESeq2::estimateSizeFactors(dds_filt)

  # Generate annotation file using biomaRt
  #FIXME:
  # Issues with Curl when using https:// and http:// throws a warning:
  #  Warning: Ensembl will soon enforce the use of https.
  #   Ensure the 'host' argument includes "https://"
  cat("... fetching annotation from biomart\n")
  ensemblAnnot <- generateEnsemblAnnotation(ensembl_ids = rownames(dds),
                                            host=biomart_host,
                                            version=biomart_Ens_version,
                                            dataset=biomart_dataset)
  
  cat("saving...", precomputed_objects_file, "\n")
  cat("...into:", output_dir, "\n")
  save(dds, dds_filt, ensemblAnnot,  file = precomputed_objects_file)
  
} else{
  cat("RNA objects file '", precomputed_objects_filename, "' exist...loading\n")
  load(precomputed_objects_file)
}

#transformation after filtering
if(!file.exists(precomputed_transf_objects_file)){
  # transformation
  log2_norm_filt <- DESeq2::normTransform(dds_filt)
  vsd_filt <- DESeq2::vst(dds_filt, blind = TRUE) # blind = TRUE for QC
  #rld_filt <- rlog(dds_filt, blind = TRUE)
  #rld_filt - too big for >100 samples; skipping!
  #rld_filt <- rlog(dds_filt, blind = FALSE) # not blind to batch effects
  save(log2_norm_filt, vsd_filt,   
       file = precomputed_transf_objects_file)
} else{
  load(precomputed_transf_objects_file)
}

```

Parameters for dataset preparation:

| Parameter           | Value                   | Decription                                               |
|---------------------|-------------------------|----------------------------------------------------------|
| experiment_name     | `r experiment_name`     | this can be a subset analysis - e.g. just batch1,...     |
| experiment_design   | `r experiment_design`   | used only to construct initial dds object                |
| abs_filt_samples    | `r abs_filt_samples`    | \# at least in N samples, which is a smallest group size |
| biomart_host        | `r biomart_host`        |                                                          |
| biomart_Ens_version | `r biomart_Ens_version` |                                                          |
| biomart_dataset     | `r biomart_dataset`     |                                                          |
|                     |                         |                                                          |


Objects overview:
```{r 03 dataset objects overview, eval=TRUE, include=TRUE, message=FALSE}
dds_objects_overview <- data.frame(dds_object = c("dds", "dds_filt"),
                                   n_samples = c(ncol(dds), ncol(dds_filt)),
                                   n_genes = c(nrow(dds), nrow(dds_filt)))

dds_objects_overview %>%
  kableExtra::kbl(caption = "dds objects overview") %>% 
  kableExtra::kable_classic(full_width = T)

# DT::datatable(dds_objects_overview,
#           options = list(pageLength = 10),
#           caption = 'Table 1: Samples in validation set')
```
#### End of the data_prep steps

```{r 03 loading more libraries - DE main, message=FALSE, eval=TRUE, include=TRUE}
#import::from(.from = SummarizedExperiment, colData, assay)
import::from(.from = variancePartition, fitExtractVarPartModel, sortCols, plotVarPart)
import::from(.from = doParallel, registerDoParallel)
import::from(.from = parallel, makeCluster, stopCluster)

# import utils scripts
import::from(.from = here::here("utils/filterDatasets.R"), "filterDatasets", .character_only=TRUE) # used for filtering
import::from(.from = here::here("utils/rnaSelectTopVarGenes.R"), "rnaSelectTopVarGenes", .character_only=TRUE)
import::from(.from = here::here("utils/edaFunctions.R"), "varPartitionEstimate", "generatePCA", "pcaExtractVariance", "pcaPlotVariance", "pcaCorrPCs", "pcaCorrPCsPlot", "generatePCA_plus_shape", .character_only=TRUE)
import::from(.from = here::here("utils/generateResults.R"), "generateResults_upd", .character_only=TRUE) # 
import::from(.from = here::here("utils/plotDegResults.R"), "plotVolcano","plotVolcano_repel", .character_only=TRUE)

library(ggrepel)

```

### Differential expression analysis

Comparison between *Tumor_plus_NK* vs *Tumor_only* - specifically *Tumor_plus_NK_timepoint_2* against *Tumor_only_timepoint_1*. *Tumor_plus_WT_NK_timepoint_2 * was renamed to *Tumor_plus_NK_timepoint_2*

Steps: 

 * subsetting dds_filt to contain only condition_tp of interest (Tumor_only_timepoint_1, Tumor_plus_NK_timepoint_2 = Tumor_plus_WT_NK_timepoint_2)
 * filter (again) for lowly expressed genes
 * results with differentially expressed genes (DEG) and significantly DEG at a selected cutoff
 * enrichment analysis (over-representation analysis) of the significantly DEG genes; GOBP, GOCC and GOMF terms enrichment

```{r 03 DE main - loading objects, message=FALSE, eval=TRUE, include=TRUE, cache=TRUE}
abs_filt_samples=3
padj_cutoff = 0.05
log2FC_cutoff = 0.58 #(FC=1.5); log2FC=1.0 # (FC=2)
var_expl_needed <- 0.6         # at least 60% variance explained needed

# compare Tumor_only between timepoints!
# condition_tp_subset <- c("Tumor_only_timepoint_0", 
#                          "Tumor_only_timepoint_1", 
#                          "Tumor_only_timepoint_2", 
#                          "Tumor_plus_NK_timepoint_1", 
#                          "Tumor_plus_NK_timepoint_2",
#                          "Tumor_plus_WT_NK_timepoint_2")

condition_tp_subset <- c("Tumor_only_timepoint_1",
                         "Tumor_plus_NK_timepoint_2",
                         "Tumor_plus_NK_timepoint_1",
                         "Tumor_plus_WT_NK_timepoint_2")

genes_of_interest <- list("Denn2b"="ENSMUSG00000031024",
                          "Gbp6"="ENSMUSG00000104713",
                          "H2-K1"="ENSMUSG00000061232",
                          "Hs3st2"="ENSMUSG00000046321",
                          "Htra3"="ENSMUSG00000029096",
                          "Ifi27"="ENSMUSG00000064215",
                          "Igtp"="ENSMUSG00000078853",
                          "Irf1"="ENSMUSG00000018899",
                          "Lgals3bp"="ENSMUSG00000033880",
                          "Ly6a"="ENSMUSG00000075602",
                          "Plaat3"="ENSMUSG00000060675",
                          "Rtp4"="ENSMUSG00000033355",
                          "Serpina3f"="ENSMUSG00000066363",
                          "Stat1"="ENSMUSG00000026104")

genes_of_interest_exp9.4_9.7 <- list(
                          'Ifi44'     ="ENSMUSG00000028037", 
                          'Ccl5'      ="ENSMUSG00000035042",
                          'Iigp1'     ="ENSMUSG00000054072",
                          'Serpina3f' ="ENSMUSG00000066363",
                          'Ly6a'      ="ENSMUSG00000075602",
                          'Plaat3'    ="ENSMUSG00000060675",
                          'Lgals3bp'  ="ENSMUSG00000033880",
                          'Ifi27'     ="ENSMUSG00000064215",
                          'Gbp6'      ="ENSMUSG00000104713",
                          'Rtp4'      ="ENSMUSG00000033355",
                          'Stat1'     ="ENSMUSG00000026104",
                          'Tap1'      ="ENSMUSG00000037321",
                          'H2-K1'     ="ENSMUSG00000061232",
                          'B2m'       ="ENSMUSG00000060802",
                          'Igtp'      ="ENSMUSG00000078853",
                          'Gtf2ird1'  ="ENSMUSG00000023079",
                          'Cdc42ep4'  ="ENSMUSG00000041598",
                          'Hoxb6'     ="ENSMUSG00000000690",
                          'Bfsp2'     ="ENSMUSG00000032556",
                          'Sox6'      ="ENSMUSG00000051910",
                          'Tgm2'      ="ENSMUSG00000037820",
                          'Adprhl1'   ="ENSMUSG00000031448")

# additional parameters ----

if(exists("deg_design")) {rm(deg_design)}
deg_design <- as.formula("~ experiment + cell_line_label + condition_tp")


# # raw and filtered dds
# precomputed_objects_filename <- paste0(experiment_name, "_dds_objects.RData")
# precomputed_objects_file <- file.path(output_dir, precomputed_objects_filename)
# 
# # log2 and vsd transformed
# precomputed_transf_objects_filename <- paste0(experiment_name, "_log2_vsd_filt_objects.RData")
# precomputed_transf_objects_file <- file.path(output_dir, precomputed_transf_objects_filename)
# 
# # check if object loaded if not load
# if(!exists(precomputed_objects_file)){
#   load(precomputed_objects_file)
#   load(precomputed_transf_objects_file)
# } 

```


```{r 03 AB DE main - preparing object - AB subset, message=FALSE, eval=TRUE, include=TRUE, cache=TRUE}
# preparing dds and vsd objects for the main project comparison 
# We use only EXP 9.4 - 9.7. EXP 9.8 is excluded.
# Preparing AB subset. 
# defining parameters

deg_name <- "TpNK_tp2_vs_Tonly_tp1_AB"
deg_dir <- file.path(output_dir, paste0("deg_", deg_name, "/"))
dir.create(deg_dir)

precomputed_objects_deg_filename <- paste0(deg_name, "_dds_objects.RData")
precomputed_objects_deg_file <- file.path(deg_dir, precomputed_objects_deg_filename)

precomputed_transf_objects_deg_filename <- paste0(deg_name, "_log2_vsd_filt_objects.RData")
precomputed_transf_objects_deg_file <- file.path(deg_dir, precomputed_transf_objects_deg_filename)

experiment_subset <- c("EXP9.4", "EXP9.5", "EXP9.6", "EXP9.7")


# just in case remove any existing subsets
if(exists("dds_subset")) {rm(dds_subset)}
if(exists("dds_subset_filt")) {rm(dds_subset_filt)}
if(exists("vsd_subset")) {rm(vsd_subset)}

if(!file.exists(precomputed_objects_deg_file)){
  cat("RNA objects file '", precomputed_objects_deg_filename, "' does not exist in the OUTPUT_DIR...\n")
  cat("generating", precomputed_objects_deg_filename, "\n")

  dds_subset <- dds_filt[ , dds_filt$condition_tp %in% condition_tp_subset]  
  dds_subset <- dds_subset[ , dds_subset$experiment %in% experiment_subset]
  dds_subset <- dds_subset[ , dds_subset$cell_line_label %in% c("A", "B")]
  
  # fixing Tumor_plus_WT_NK_timepoint_2 -> Tumor_plus_NK_timepoint_2
  dds_subset$condition_tp <- droplevels(dds_subset$condition_tp)
  table(dds_subset$condition_tp)
  
  dds_subset$experiment <- droplevels(dds_subset$experiment)
  table(dds_subset$experiment)
  
  dds_subset$cell_line_label <- droplevels(dds_subset$cell_line_label)
  table(dds_subset$cell_line_label)
  
  # filtering lowly expressed genes
  dds_subset_filt <- filterDatasets(dds_subset, 
                                    abs_filt = TRUE, 
                                    abs_filt_samples = abs_filt_samples) # at least in N samples, which is a smallest group size

  cat("... dds\n")
  design(dds_subset_filt) <- deg_design
  dds_subset_filt <- DESeq2::estimateSizeFactors(dds_subset_filt)
  dds_subset_filt <- DESeq2::DESeq(dds_subset_filt) # do not replace outliers based on replicates
  
  cat("saving...", precomputed_objects_file, "\n")
  cat("...into:", output_dir,deg_name,"\n")
  save(dds_subset, dds_subset_filt, ensemblAnnot,  file = precomputed_objects_deg_file)
  
} else{
  cat("RNA objects file '", precomputed_objects_deg_filename, "' exist...loading\n")
  load(precomputed_objects_deg_file)
}

#transformation after filtering
if(!file.exists(precomputed_transf_objects_deg_file)){
  # transformation
  log2_norm_subset_filt <- DESeq2::normTransform(dds_subset_filt)
  vsd_subset_filt <- DESeq2::vst(dds_subset_filt, blind = FALSE) # using blind=FALSE utilize design info; blind = TRUE for QC
  #rld_filt <- DESeq2::rlog(dds_subset_filt, blind = FALSE) # more robust to differences in sequencing depth!
  #rld_filt - too big for >100 samples; skipping!
  #rld_filt <- rlog(dds_filt, blind = FALSE) # not blind to batch effects
  save(log2_norm_subset_filt, vsd_subset_filt,   
       file = precomputed_transf_objects_deg_file)
} else{
  load(precomputed_transf_objects_deg_file)
}
```


```{r 03 AB "composing the raw count tables for Ab3 DiffTF project"}

#get the counts table for diffTFs  AB cell lines
fl_smpl_nm <- gsub("_Day\\d\\d","",dds_subset@colData@listData[["full_sample_name"]])
fl_smpl_nm <- gsub("_\\d_","_",fl_smpl_nm)

sample_name <- paste0(dds_subset@colData@listData[["original_experiment"]], "_",
                      fl_smpl_nm, "_",
                      dds_subset@colData@listData[["day_cell_harvesting"]], "_REP",
                      dds_subset@colData@listData[["technical_replicate"]])

sample_name <- gsub("^EXP\\s9.", "EXP9_", sample_name)
sample_name <- gsub("\\+", "_plus_", sample_name)
sample_name <- gsub("_Tumor_only_Day\\d_", "_Tumor_only_", sample_name)

#These samples from RNA-seq are missing in ATAC data.
#EXP9_6_13G_Tumor_plus_NK_Day14_REP1 is missing in ./
#EXP9_6_13G_Tumor_plus_NK_Day14_REP2 is missing in ./
#EXP9_6_13G_Tumor_plus_NK_Day14_REP3 is missing in ./
#EXP9_6_13H_Tumor_plus_NK_Day14_REP1 is missing in ./
#EXP9_6_13H_Tumor_plus_NK_Day14_REP2 is missing in ./
#EXP9_6_13H_Tumor_plus_NK_Day14_REP3 is missing in ./
#
#Instead there is Day_10 version.
#
#I'm going to replace in the RNA-seq count table the day 14 with day 10, so that samples match.

sample_name <- gsub("EXP9_6_13G_Tumor_plus_NK_Day14", "EXP9_6_13G_Tumor_plus_NK_Day10", sample_name)
sample_name <- gsub("EXP9_6_13H_Tumor_plus_NK_Day14", "EXP9_6_13H_Tumor_plus_NK_Day10", sample_name)
col_IDS <- c("ENSEMBL", sample_name)
RNA_seq_raw_counts <- dds_subset@assays@data@listData[["counts"]]
RNA_seq_raw_counts <- cbind(row.names(RNA_seq_raw_counts), RNA_seq_raw_counts)
RNA_seq_raw_counts <- rbind(col_IDS,RNA_seq_raw_counts)

write.table(RNA_seq_raw_counts, file = "~/workspace/RNA_seq_raw_counts_AB.tsv",
            sep ="\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

# Building the sample table
sampleID_list <- list("sampleID" = sample_name)
bamReads_list <- list("bamReads" = paste0("/home/aleksandr_b/bioinf_isilon/core_bioinformatics_unit/Internal/aleksandr/projects/ab3_20230220_nk_tumor_project/data_ATACSeq_nfcore_run/results/bwa/merged_library/", sample_name, ".mLb.clN.sorted.bam"))
cond_summ <- sample_name
cond_summ[stringr::str_detect(cond_summ, pattern = "_Tumor_only_")] <- "Tumor_only_TP_1"
cond_summ[stringr::str_detect(cond_summ, pattern = "_Tumor_plus_NK")] <- "Tumor_pls_NK_TP2"
conditionSummary_list <- list("conditionSummary" = cond_summ)
experiment_id_list <- list("experiment_id" = stringr::str_extract(sample_name, "EXP\\d_\\d"))

sampleData.tsv <- data.frame(sampleID_list,
                             bamReads_list,
                             conditionSummary_list,
                             experiment_id_list)
write.table(sampleData.tsv, file = "~/workspace/sampleData.tsv",
            sep ="\t", quote = FALSE, row.names = FALSE)

```


```{r 03 AB DE main - key variables table, message=FALSE, eval=TRUE, include=TRUE}
#print(key_variables_tableOne$CatTable)
rm(key_metadata_subset, key_variables_tableOne) # in case this exists from previous runs
key_metadata_subset <- as.data.frame(colData(dds_subset)) %>%
  dplyr::select(experiment, condition_tp, cell_line_label, technical_replicate) 

key_variables_tableOne <- tableone::CreateTableOne(vars = colnames(dplyr::select(key_metadata_subset, -condition_tp)), 
                           strata = c("condition_tp"), 
                           data = key_metadata_subset)

tableone::kableone(key_variables_tableOne$CatTable,
                   caption = "Overview of number of samples in different categories (experiment, condition,...).") %>%
  kableExtra::kable_material(c("striped", "hover"))
```

```{r 03 DE main - EDA _ AB Subset, message=FALSE, eval=TRUE, include=TRUE, cache=TRUE}
#TODO:
# - plot images after removing batch effects!
# EDA and visualize plots before and after batch effect correction

# transf_object <- vsd_subset_filt
# pca_deg <- generatePCA(transf_object = transf_object, 
#                        cond_interest_varPart = c("condition_tp", "cell_line_label", "experiment"), 
#                        color_variable = "condition_tp", 
#                        shape_variable = "experiment",
#                        ntop_genes = 1000) +
#   ggtitle("Original dataset") +
#  geom_text_repel(aes(label = transf_object$sample_name),
#                   box.padding = 3)
# 
# 
# 
# pca_deg_plotCellLabel <- generatePCA(transf_object = transf_object, 
#                        cond_interest_varPart = c("condition_tp", "cell_line_label", "experiment"), 
#                        color_variable = "condition_tp", 
#                        shape_variable = "cell_line_label",
#                        ntop_genes = 1000) +
#   ggtitle("Original dataset")
# 
# # Remove batch effect
# transf_batch_NObatch_experiment <- vsd_subset_filt
# transf_batch_NObatch_experiment_count <- limma::removeBatchEffect(SummarizedExperiment::assay(transf_batch_NObatch_experiment), transf_batch_NObatch_experiment$experiment)
# SummarizedExperiment::assay(transf_batch_NObatch_experiment) <- transf_batch_NObatch_experiment_count
# 
# pca_deg_NObatch_experiment <- generatePCA(transf_object = transf_batch_NObatch_experiment, 
#                        cond_interest_varPart = c("condition_tp", "cell_line_label", "experiment"), 
#                        color_variable = "condition_tp", 
#                        shape_variable = "experiment",
#                        ntop_genes = 1000) +
#   ggtitle("Removed batchEffect - experiment")
# 
# pca_deg_NObatch_experiment_plotCellLabel <- generatePCA(transf_object = transf_batch_NObatch_experiment, 
#                        cond_interest_varPart = c("condition_tp", "cell_line_label", "experiment"), 
#                        color_variable = "condition_tp", 
#                        shape_variable = "cell_line_label",
#                        ntop_genes = 1000) +
#   ggtitle("Removed batchEffect - experiment")
# 
# #PCA_batch_merged <- ggpubr::ggarrange(pca_deg, pca_deg_NObatch_experiment, common.legend = TRUE)
# 
# 
# PCA_batch_merged <- ggpubr::ggarrange(pca_deg, 
#                                       pca_deg_NObatch_experiment,
#                                       common.legend = TRUE)
# 
# PCA_batch_highlightCellLines_merged <- ggpubr::ggarrange(pca_deg_plotCellLabel, 
#                                       pca_deg_NObatch_experiment_plotCellLabel,
#                                       common.legend = TRUE)


###### Plots for publication #######
transf_object <- vsd_subset_filt

pca_deg <- generatePCA_plus_shape(transf_object = transf_object, 
                       cond_interest_varPart = c("cell_line_label", "experiment","condition_tp"), 
                       color_variable = "condition_tp", 
                       shape_variable = "Cell_line_label_and_exp_id",
                       ntop_genes = 1000)+
  ggtitle("Original dataset") +
  scale_shape_manual(values = c(0,1,2,15,16,17)) +
  scale_color_manual(values = c("#DD3344", "#FF9F1C" ,"#553388"))
pca_deg

# remove batch effect
transf_batch_NObatch_experiment <- vsd_subset_filt
transf_batch_NObatch_experiment_count <- limma::removeBatchEffect(SummarizedExperiment::assay(transf_batch_NObatch_experiment), transf_batch_NObatch_experiment$experiment)
SummarizedExperiment::assay(transf_batch_NObatch_experiment) <- transf_batch_NObatch_experiment_count


pca_deg_NObatch_experiment <- generatePCA_plus_shape(transf_object = transf_batch_NObatch_experiment, 
                       cond_interest_varPart = c("cell_line_label", "experiment","condition_tp"), 
                       color_variable = "condition_tp", 
                       shape_variable = "Cell_line_label_and_exp_id",
                       ntop_genes = 1000)+
  ggtitle("Batch Corrected Dataset") +
  scale_shape_manual(values = c(0,1,2,15,16,17)) +
  scale_color_manual(values = c("#DD3344", "#FF9F1C" ,"#553388"))
pca_deg_NObatch_experiment


ggsave(filename = paste0(deg_dir,"publication/", "TpNK_tp2_vs_Tonly_tp1_AB_PCA_not_batch_corrected.png"), 
       plot=pca_deg,
       width = 20, height = 14, units = "cm")

ggsave(filename = paste0(deg_dir,"publication/", "TpNK_tp2_vs_Tonly_tp1_AB_PCA_batch_corrected.png"), 
       plot=pca_deg_NObatch_experiment,
       width = 20, height = 14, units = "cm")

ggsave(filename = paste0(deg_dir,"publication/", "TpNK_tp2_vs_Tonly_tp1_AB_PCA_not_batch_corrected.eps"), 
       plot=pca_deg,
       width = 20, height = 14, units = "cm")

ggsave(filename = paste0(deg_dir,"publication/", "TpNK_tp2_vs_Tonly_tp1_AB_PCA_batch_corrected.eps"), 
       plot=pca_deg_NObatch_experiment,
       width = 20, height = 14, units = "cm")

```

The following PCA plots show effect of removing batch effects (experiment). There are two weird samples from EXP9.7 that create a lot of variance. What should i do with them?


```{r 03 AB 2 DE main - PCA batch effect removal - experiment + cell_linet, message=FALSE, eval=TRUE, include=TRUE, fig.dim = c(12, 12)}
print(pca_deg_NObatch_experiment)
print(pca_deg)
```

```{r 03 AB DE main - DEG, message=FALSE, eval=TRUE, include=TRUE, cache=TRUE}
#resultsNames(dds_subset_filt)
# timepoint2 (T+NK)_vs_timepoint1 (T)
deg_TpNK_tp2_vs_Tonly_tp1_results <- generateResults_upd(dds_object = dds_subset_filt, 
                                                             coeff_name = "condition_tp_Tumor_plus_NK_timepoint_2_vs_Tumor_only_timepoint_1",
                                                             cond_numerator = "Tumor_plus_NK_timepoint_2", 
                                                             cond_denominator = "Tumor_only_timepoint_1",
                                                             cond_variable="condition_tp")

openxlsx::write.xlsx(deg_TpNK_tp2_vs_Tonly_tp1_results, file = file.path(deg_dir, "deg_TpNK_tp2_vs_Tonly_tp1_AB_results.xlsx"))

AB_results <- deg_TpNK_tp2_vs_Tonly_tp1_results
```

```{r 03 AB DE main - enrichment, message=FALSE, eval=TRUE, include=TRUE, cache=TRUE}
# Loading MsigDB geneset collections ----
gs_hallmark <- hypeR::msigdb_gsets(species = "Mus musculus", category = c("H"), clean=TRUE) 
gs_C2_kegg <- hypeR::msigdb_gsets(species = "Mus musculus", category = c("C2"), subcategory = "CP:KEGG", clean=TRUE) 
gs_C5_GOBP <- hypeR::msigdb_gsets(species = "Mus musculus", category = c("C5"), subcategory = "GO:BP", clean=TRUE) 
gs_C5_GOCC <- hypeR::msigdb_gsets(species = "Mus musculus", category = c("C5"), subcategory = "GO:CC", clean=TRUE) 
gs_C5_GOMF <- hypeR::msigdb_gsets(species = "Mus musculus", category = c("C5"), subcategory = "GO:MF", clean=TRUE) 


TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOBP <- hypeR::hypeR(signature = deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$mgi_symbol, 
                                                      genesets = gs_C5_GOBP, 
                                                      test="hypergeometric", 
                                                      background=nrow(dds_subset_filt))
AB_GOBP <- TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOBP

TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOCC <- hypeR::hypeR(signature = deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$mgi_symbol, 
                                                      genesets = gs_C5_GOCC, 
                                                      test="hypergeometric", 
                                                      background=nrow(dds_subset_filt))
AB_GOCC <- TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOCC

TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOMF <- hypeR::hypeR(signature = deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$mgi_symbol, 
                                                      genesets = gs_C5_GOMF, 
                                                      test="hypergeometric", 
                                                      background=nrow(dds_subset_filt))
AB_GOMF <- TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOMF


#hypeR::hyp_show(hyp_obj$data$clA, simple = FALSE)
hypeR::hyp_to_excel(TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOBP, file_path=file.path(deg_dir, "TpNK_tp2_vs_Tonly_tp1_AB_enrichment_C5_GOBP.xlsx"))
hypeR::hyp_to_excel(TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOCC, file_path=file.path(deg_dir, "TpNK_tp2_vs_Tonly_tp1_AB_enrichment_C5_GOCC.xlsx"))
hypeR::hyp_to_excel(TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOMF, file_path=file.path(deg_dir, "TpNK_tp2_vs_Tonly_tp1_AB_enrichment_C5_GOMF.xlsx"))

# plotting enrichment results ----
TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOBP_plot <- hypeR::hyp_dots(TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOBP, merge=TRUE, fdr=0.05, top = 20, abrv=70, val="fdr", title="GOBP: T+NK timepoint2 vs T-only timepoint1 AB") +theme_bw()
TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOCC_plot <- hypeR::hyp_dots(TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOCC, merge=TRUE, fdr=0.05, top = 20, abrv=70, val="fdr", title="GOCC: T+NK timepoint2 vs T-only timepoint1 AB") +theme_bw()
TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOMF_plot <- hypeR::hyp_dots(TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOMF, merge=TRUE, fdr=0.05, top = 20, abrv=70, val="fdr", title="GOMF:T+NK timepoint2 vs T-only timepoint1 AB") + theme_bw()


###### For publication ######
ggsave(filename = paste0(deg_dir,"publication/" ,"TpNK_tp2_vs_Tonly_tp1_AB_hyp_GOBP_plot.png"), 
       plot=TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOBP_plot,
       width = 20, height = 20, units = "cm")
ggsave(filename = paste0(deg_dir,"publication/" ,"TpNK_tp2_vs_Tonly_tp1_AB_hyp_GOBP_plot.eps"), 
       plot=TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOBP_plot,
       width = 20, height = 20, units = "cm")
```

### Highlighting genes of interest

Volcano plot highlighting genes of interest:
```{r 03 AB DE main - plotting results - volcano plot, message=FALSE, eval=TRUE, include=TRUE, cache=TRUE}
# volcano plot
# heatmap - corrected
genes_of_interest
genes_of_interest_exp9.4_9.7


TpNK_tp2_vs_Tonly_tp1_volcano_plot_1 <- plotVolcano(dds_results_obj = deg_TpNK_tp2_vs_Tonly_tp1_results$results_all, 
                                             genes_of_interest = names(genes_of_interest), 
                                             plot_title = "T+NK tp2 vs T-only tp1 AB genes set 1")

TpNK_tp2_vs_Tonly_tp1_volcano_plot_2 <- plotVolcano(dds_results_obj = deg_TpNK_tp2_vs_Tonly_tp1_results$results_all, 
                                             genes_of_interest = names(genes_of_interest_exp9.4_9.7), 
                                             plot_title = "T+NK tp2 vs T-only tp1 AB genes set 2")


##### For publication #####

TpNK_tp2_vs_Tonly_tp1_volcano_plot_2 <- plotVolcano_repel(dds_results_obj = deg_TpNK_tp2_vs_Tonly_tp1_results$results_all, 
                                             genes_of_interest = names(genes_of_interest_exp9.4_9.7), 
                                             plot_title = "T+NK tp2 vs T-only tp1 AB")

ggsave(filename = paste0(deg_dir,"publication/", "TpNK_tp2_vs_Tonly_tp1_AB_volcano_plot.png"), 
       plot=TpNK_tp2_vs_Tonly_tp1_volcano_plot_2,
       width = 18, height = 12, units = "cm")
ggsave(filename = paste0(deg_dir,"publication/", "TpNK_tp2_vs_Tonly_tp1_AB_volcano_plot.eps"), 
       plot=TpNK_tp2_vs_Tonly_tp1_volcano_plot_2,
       width = 18, height = 12, units = "cm")
```

```{r 03 AB 2 DE main - plotting results - volcano plot print, message=FALSE, eval=TRUE, include=TRUE, fig.dim = c(10, 10)}
print(TpNK_tp2_vs_Tonly_tp1_volcano_plot_1)
print(TpNK_tp2_vs_Tonly_tp1_volcano_plot_2)

```

Heatmap of normalized and scaled expression of `r length(deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$ensembl_id)` significantly differentially expressed genes before and after removeal of batch effects coming from different experiments:

```{r 03 AB DE main - plotting results - heatmap plot, message=FALSE, eval=TRUE, include=TRUE, cache=TRUE, fig.dim = c(12, 14)}
# heatmap and heatmap batch corrected - corrected 
metadata_heatmap <- as.data.frame(colData(dds_subset))

heatmap_counts <- SummarizedExperiment::assay(transf_object)
heatmap_counts_NObatch <- SummarizedExperiment::assay(transf_batch_NObatch_experiment) # removed xperiment batch effects! ? use experimen+cell_line removed???

heatmap_counts_deg <- heatmap_counts[rownames(heatmap_counts) %in% deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$ensembl_id, ]
heatmap_counts_NObatch_deg <- heatmap_counts_NObatch[rownames(heatmap_counts_NObatch) %in% deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$ensembl_id, ]

annotation_col <- metadata_heatmap %>%
  dplyr::select(experiment, cell_line_label, condition_tp) %>% # condition, timepoint_cell_harvesting, 
  dplyr::arrange(condition_tp, cell_line_label)

heatmap_counts_deg_ord <- heatmap_counts_deg[, match(rownames(annotation_col), colnames(heatmap_counts_deg))]
heatmap_counts_NObatch_deg_ord <- heatmap_counts_NObatch_deg[, match(rownames(annotation_col), colnames(heatmap_counts_NObatch_deg))]

ensembl2symbol_annot <- deg_TpNK_tp2_vs_Tonly_tp1_results$results_all %>%
  dplyr::select(ensembl_id, mgi_symbol)

color.scheme <- rev(RColorBrewer::brewer.pal(8,"RdBu")) # generate the color scheme to use

ann_colors = list(
  experiment = c( EXP9.5 = "#005f73", EXP9.6 = "#0a9396", EXP9.7 = "#94d2bd"),
  condition_tp = c(Tumor_only_timepoint_1 = "#DD3344", Tumor_plus_NK_timepoint_1 = "#FF9F1C",  Tumor_plus_NK_timepoint_2 = "#553388"),
  cell_line_label = c(A = "#E9D8A6", B = "#D9BE6D")
)

heatmap_not_corrected <- pheatmap::pheatmap(heatmap_counts_deg_ord,
                   main = "Heatmap of signif. DEG",
                   scale = "row",
                   annotation_col = annotation_col,
                   annotation_colors = ann_colors,
                   #annotation_row = row_annot_symbols,
                   show_colnames = FALSE,
                   show_rownames = FALSE,
                   cluster_cols = FALSE,
                   #cluster_rows = counts_deg_ord_row_cor_hclust,
                   color = color.scheme,
                   fontsize = 10, fontsize_row = 10 #height=10, cellwidth = 11, cellheight = 11
                   ) # 

ggsave(filename = paste0(deg_dir, "TpNK_tp2_vs_Tonly_tp1_AB_heatmap_not_corrected.png"), 
       plot=heatmap_not_corrected,
       width = 20, height = 20, units = "cm")


```

```{r 03 AB DE main - plotting results - heatmap plot - batchEffect experiment removed, message=FALSE, eval=TRUE, include=TRUE, cache=TRUE, fig.dim = c(12, 14)}
# heatmap and heatmap batch corrected - corrected 
metadata_heatmap <- as.data.frame(colData(dds_subset))

heatmap_counts <- SummarizedExperiment::assay(transf_object)
heatmap_counts_NObatch <- SummarizedExperiment::assay(transf_batch_NObatch_experiment) # removed xperiment batch effects! ? use experimen+cell_line removed???

heatmap_counts_deg <- heatmap_counts[rownames(heatmap_counts) %in% deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$ensembl_id, ]
heatmap_counts_NObatch_deg <- heatmap_counts_NObatch[rownames(heatmap_counts_NObatch) %in% deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$ensembl_id, ]

annotation_col <- metadata_heatmap %>%
  dplyr::select(experiment, cell_line_label, condition_tp) %>% # condition, timepoint_cell_harvesting, 
  dplyr::arrange(condition_tp, cell_line_label)

heatmap_counts_deg_ord <- heatmap_counts_deg[, match(rownames(annotation_col), colnames(heatmap_counts_deg))]
heatmap_counts_NObatch_deg_ord <- heatmap_counts_NObatch_deg[, match(rownames(annotation_col), colnames(heatmap_counts_NObatch_deg))]

ensembl2symbol_annot <- deg_TpNK_tp2_vs_Tonly_tp1_results$results_all %>%
  dplyr::select(ensembl_id, mgi_symbol)

color.scheme <- rev(RColorBrewer::brewer.pal(8,"RdBu")) # generate the color scheme to use

ann_colors = list(
  experiment = c( EXP9.5 = "#005f73", EXP9.6 = "#0a9396", EXP9.7 = "#94d2bd"),
  condition_tp = c(Tumor_only_timepoint_1 = "#DD3344", Tumor_plus_NK_timepoint_1 = "#FF9F1C",  Tumor_plus_NK_timepoint_2 = "#553388"),
  cell_line_label = c(A = "#E9D8A6", B = "#D9BE6D")
)

heatmap_corrected <- pheatmap::pheatmap(heatmap_counts_NObatch_deg_ord,
                   main = "Heatmap of signif. DEG - Batch Corrected",
                   scale = "row",
                   annotation_col = annotation_col,
                   annotation_colors = ann_colors,
                   #annotation_row = row_annot_symbols,
                   show_colnames = FALSE,
                   show_rownames = FALSE,
                   cluster_cols = FALSE,
                   #cluster_rows = counts_deg_ord_row_cor_hclust,
                   color = color.scheme,
                   fontsize = 10, fontsize_row = 10 #height=10, cellwidth = 11, cellheight = 11
                   ) # 

ggsave(filename = paste0(deg_dir, "TpNK_tp2_vs_Tonly_tp1_AB_heatmap_corrected.png"), 
       plot=heatmap_corrected,
       width = 20, height = 20, units = "cm")


##### For publication #####
ggsave(filename = paste0(deg_dir, "publication/", "TpNK_tp2_vs_Tonly_tp1_AB_heatmap_corrected.eps"), 
       plot=heatmap_corrected,
       width = 18, height = 20, units = "cm")
ggsave(filename = paste0(deg_dir, "publication/", "TpNK_tp2_vs_Tonly_tp1_AB_heatmap_corrected.png"), 
       plot=heatmap_corrected,
       width = 18, height = 20, units = "cm")
```

```{r 03 AB "small heatmap"}
# identify the 25 gene set
minL2FC_Ids <- order(deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$log2FoldChange)[1:25]
maxL2FC_Ids <- order(decreasing = T, deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$log2FoldChange)[1:25]
L2FC_EMSEMBL <- deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$ensembl_id[c(minL2FC_Ids, maxL2FC_Ids)]
L2FC_EMSEMBL <- c(L2FC_EMSEMBL, "ENSMUSG00000075602", "ENSMUSG00000060675")

heatmap_counts_NObatch <- SummarizedExperiment::assay(transf_batch_NObatch_experiment) # removed xperiment batch effects! ? use experimen+cell_line removed???

heatmap_counts_NObatch_deg <- heatmap_counts_NObatch[rownames(heatmap_counts_NObatch) %in% L2FC_EMSEMBL, ]

annotation_col <- metadata_heatmap %>%
  dplyr::select(experiment, cell_line_label, condition_tp) %>% # condition, timepoint_cell_harvesting, 
  dplyr::arrange(condition_tp, cell_line_label)

heatmap_counts_NObatch_deg_ord <- heatmap_counts_NObatch_deg[, match(rownames(annotation_col), colnames(heatmap_counts_NObatch_deg))]

ensembl2symbol_annot <- deg_TpNK_tp2_vs_Tonly_tp1_results$results_all %>%
  dplyr::select(ensembl_id, mgi_symbol)

rownames(heatmap_counts_NObatch_deg_ord) <- ensembl2symbol_annot[match(x = row.names(heatmap_counts_NObatch_deg_ord), table = ensembl2symbol_annot$ensembl_id),]$mgi_symbol

color.scheme <- rev(RColorBrewer::brewer.pal(8,"RdBu")) # generate the color scheme to use

ann_colors = list(
  experiment = c( EXP9.5 = "#005f73", EXP9.6 = "#0a9396", EXP9.7 = "#94d2bd"),
  condition_tp = c(Tumor_only_timepoint_1 = "#DD3344", Tumor_plus_NK_timepoint_1 = "#FF9F1C",  Tumor_plus_NK_timepoint_2 = "#553388"),
  cell_line_label = c(A = "#E9D8A6", B = "#D9BE6D")
)


heatmap_corrected <- pheatmap::pheatmap(heatmap_counts_NObatch_deg_ord,
                   main = "Heatmap of signif. DEG - Batch Corrected 50 most regulated genes",
                   scale = "row",
                   annotation_col = annotation_col,
                   annotation_colors = ann_colors,
                   #annotation_row = row_annot_symbols,
                   show_colnames = FALSE,
                   show_rownames = T,
                   cluster_cols = FALSE,
                   #cluster_rows = counts_deg_ord_row_cor_hclust,
                   color = color.scheme,
                   fontsize = 10, fontsize_row = 10 #height=10, cellwidth = 11, cellheight = 11
                   ) # 

##### For publication #####
ggsave(filename = paste0(deg_dir, "publication/", "TpNK_tp2_vs_Tonly_tp1_AB_heatmap_corrected_50genes.eps"), 
       plot=heatmap_corrected,
       width = 24, height = 20, units = "cm")
ggsave(filename = paste0(deg_dir, "publication/", "TpNK_tp2_vs_Tonly_tp1_AB_heatmap_corrected_50_genes.png"), 
       plot=heatmap_corrected,
       width = 24, height = 20, units = "cm")
```

### Now - process the CD subset. Later I will check the overlap between these two.
```{r 03  DE main - preparing object - CD subset, message=FALSE, eval=TRUE, include=TRUE, cache=TRUE}
# preparing dds and vsd objects for the main project comparison 
# We use only EXP 9.4 - 9.7. EXP 9.8 is excluded.
# Preparing CD subset. 
# defining parameters

deg_name <- "TpNK_tp2_vs_Tonly_tp1_CD"
deg_dir <- file.path(output_dir, paste0("deg_", deg_name, "/"))
dir.create(deg_dir)

precomputed_objects_deg_filename <- paste0(deg_name, "_dds_objects.RData")
precomputed_objects_deg_file <- file.path(deg_dir, precomputed_objects_deg_filename)

precomputed_transf_objects_deg_filename <- paste0(deg_name, "_log2_vsd_filt_objects.RData")
precomputed_transf_objects_deg_file <- file.path(deg_dir, precomputed_transf_objects_deg_filename)

experiment_subset <- c("EXP9.4", "EXP9.5", "EXP9.6", "EXP9.7")


# just in case remove any existing subsets
if(exists("dds_subset")) {rm(dds_subset)}
if(exists("dds_subset_filt")) {rm(dds_subset_filt)}
if(exists("vsd_subset")) {rm(vsd_subset)}

if(!file.exists(precomputed_objects_deg_file)){
  cat("RNA objects file '", precomputed_objects_deg_filename, "' does not exist in the OUTPUT_DIR...\n")
  cat("generating", precomputed_objects_deg_filename, "\n")

  dds_subset <- dds_filt[ , dds_filt$condition_tp %in% condition_tp_subset]  
  dds_subset <- dds_subset[ , dds_subset$experiment %in% experiment_subset]
  dds_subset <- dds_subset[ , dds_subset$cell_line_label %in% c("C", "D")]
  
  # fixing Tumor_plus_WT_NK_timepoint_2 -> Tumor_plus_NK_timepoint_2
  dds_subset$condition_tp <- droplevels(dds_subset$condition_tp)
  table(dds_subset$condition_tp)
  
  dds_subset$experiment <- droplevels(dds_subset$experiment)
  table(dds_subset$experiment)
  
  dds_subset$cell_line_label <- droplevels(dds_subset$cell_line_label)
  table(dds_subset$cell_line_label)
  
  # filtering lowly expressed genes
 
  dds_subset_filt <- filterDatasets(dds_subset, 
                                    abs_filt = TRUE, 
                                    abs_filt_samples = abs_filt_samples) # at least in N samples, which is a smallest group size

  cat("... dds\n")
  design(dds_subset_filt) <- deg_design
  dds_subset_filt <- DESeq2::estimateSizeFactors(dds_subset_filt)
  dds_subset_filt <- DESeq2::DESeq(dds_subset_filt) # do not replace outliers based on replicates
  
  cat("saving...", precomputed_objects_file, "\n")
  cat("...into:", output_dir, "\n")
  save(dds_subset, dds_subset_filt, ensemblAnnot,  file = precomputed_objects_deg_file)
  
} else{
  cat("RNA objects file '", precomputed_objects_deg_filename, "' exist...loading\n")
  load(precomputed_objects_deg_file)
}

#transformation after filtering
if(!file.exists(precomputed_transf_objects_deg_file)){
  # transformation
  log2_norm_subset_filt <- DESeq2::normTransform(dds_subset_filt)
  vsd_subset_filt <- DESeq2::vst(dds_subset_filt, blind = FALSE) # using blind=FALSE utilize design info; blind = TRUE for QC
  #rld_filt <- DESeq2::rlog(dds_subset_filt, blind = FALSE) # more robust to differences in sequencing depth!
  #rld_filt - too big for >100 samples; skipping!
  #rld_filt <- rlog(dds_filt, blind = FALSE) # not blind to batch effects
  save(log2_norm_subset_filt, vsd_subset_filt,   
       file = precomputed_transf_objects_deg_file)
} else{
  load(precomputed_transf_objects_deg_file)
}

```
```{r 03 CD DE main - key variables table, message=FALSE, eval=TRUE, include=TRUE}
#print(key_variables_tableOne$CatTable)
rm(key_metadata_subset, key_variables_tableOne) # in case this exists from previus runs
key_metadata_subset <- as.data.frame(colData(dds_subset)) %>%
  dplyr::select(experiment, condition_tp, cell_line_label, technical_replicate) 

key_variables_tableOne <- tableone::CreateTableOne(vars = colnames(dplyr::select(key_metadata_subset, -condition_tp)), 
                           strata = c("condition_tp"), 
                           data = key_metadata_subset)

tableone::kableone(key_variables_tableOne$CatTable,
                   caption = "Overview of number of samples in different categories (experiment, condition,...).") %>%
  kableExtra::kable_material(c("striped", "hover"))
```

```{r 03 CD DE main - EDA Subset, message=FALSE, eval=TRUE, include=TRUE, cache=TRUE}
#TODO:
# - plot images after removing batch effects!
# EDA and visualize plots before and after batch effect correction
transf_object <- vsd_subset_filt
pca_deg <- generatePCA(transf_object = transf_object, 
                       cond_interest_varPart = c("condition_tp", "cell_line_label", "experiment"), 
                       color_variable = "condition_tp", 
                       shape_variable = "experiment",
                       ntop_genes = 1000) +
  ggtitle("Original dataset")

pca_deg_plotCellLabel <- generatePCA(transf_object = transf_object, 
                       cond_interest_varPart = c("condition_tp", "cell_line_label", "experiment"), 
                       color_variable = "condition_tp", 
                       shape_variable = "cell_line_label",
                       ntop_genes = 1000) +
  ggtitle("Original dataset")

# Remove batch effect
# remove batch effect
transf_batch_NObatch_experiment <- vsd_subset_filt
transf_batch_NObatch_experiment_count <- limma::removeBatchEffect(SummarizedExperiment::assay(transf_batch_NObatch_experiment), transf_batch_NObatch_experiment$experiment)
SummarizedExperiment::assay(transf_batch_NObatch_experiment) <- transf_batch_NObatch_experiment_count

pca_deg_NObatch_experiment <- generatePCA(transf_object = transf_batch_NObatch_experiment, 
                       cond_interest_varPart = c("condition_tp", "cell_line_label", "experiment"), 
                       color_variable = "condition_tp", 
                       shape_variable = "experiment",
                       ntop_genes = 1000) +
  ggtitle("Removed batchEffect - experiment")

pca_deg_NObatch_experiment_plotCellLabel <- generatePCA(transf_object = transf_batch_NObatch_experiment, 
                       cond_interest_varPart = c("condition_tp", "cell_line_label", "experiment"), 
                       color_variable = "condition_tp", 
                       shape_variable = "cell_line_label",
                       ntop_genes = 1000) +
  ggtitle("Removed batchEffect - experiment")

#PCA_batch_merged <- ggpubr::ggarrange(pca_deg, pca_deg_NObatch_experiment, common.legend = TRUE)


PCA_batch_merged <- ggpubr::ggarrange(pca_deg, 
                                      pca_deg_NObatch_experiment,
                                      common.legend = TRUE)

PCA_batch_highlightCellLines_merged <- ggpubr::ggarrange(pca_deg_plotCellLabel, 
                                      pca_deg_NObatch_experiment_plotCellLabel,
                                      common.legend = TRUE)


#### for publication #####
transf_object <- vsd_subset_filt
pca_deg <- generatePCA_plus_shape(transf_object = transf_object, 
                       cond_interest_varPart = c("cell_line_label", "experiment","condition_tp"), 
                       color_variable = "condition_tp", 
                       shape_variable = "Cell_line_label_and_exp_id",
                       ntop_genes = 1000)+
  ggtitle("Original dataset") +
  scale_shape_manual(values = c(0,1,2,5,15,16,17,18)) +
  scale_color_manual(values = c("#DD3344", "#FF9F1C", "#553388"))
pca_deg

# remove batch effect
transf_batch_NObatch_experiment <- vsd_subset_filt
transf_batch_NObatch_experiment_count <- limma::removeBatchEffect(SummarizedExperiment::assay(transf_batch_NObatch_experiment), transf_batch_NObatch_experiment$experiment)
SummarizedExperiment::assay(transf_batch_NObatch_experiment) <- transf_batch_NObatch_experiment_count


pca_deg_NObatch_experiment <- generatePCA_plus_shape(transf_object = transf_batch_NObatch_experiment, 
                       cond_interest_varPart = c("cell_line_label", "experiment","condition_tp"), 
                       color_variable = "condition_tp", 
                       shape_variable = "Cell_line_label_and_exp_id",
                       ntop_genes = 1000)+
  ggtitle("Batch Corrected Dataset") +
  scale_shape_manual(values = c(0,1,2,5,15,16,17,18)) +
  scale_color_manual(values = c("#DD3344", "#FF9F1C", "#553388"))
pca_deg_NObatch_experiment


ggsave(filename = paste0(deg_dir,"publication/", "TpNK_tp2_vs_Tonly_tp1_CD_PCA_not_batch_corrected.png"), 
       plot=pca_deg,
       width = 20, height = 14, units = "cm")

ggsave(filename = paste0(deg_dir,"publication/", "TpNK_tp2_vs_Tonly_tp1_CD_PCA_batch_corrected.png"), 
       plot=pca_deg_NObatch_experiment,
       width = 20, height = 14, units = "cm")

ggsave(filename = paste0(deg_dir,"publication/", "TpNK_tp2_vs_Tonly_tp1_CD_PCA_not_batch_corrected.eps"), 
       plot=pca_deg,
       width = 20, height = 14, units = "cm")

ggsave(filename = paste0(deg_dir,"publication/", "TpNK_tp2_vs_Tonly_tp1_CD_PCA_batch_corrected.eps"), 
       plot=pca_deg_NObatch_experiment,
       width = 20, height = 14, units = "cm")


```


```{r 03 CD 1 DE main - DEG, message=FALSE, eval=TRUE, include=TRUE, cache=TRUE}
#resultsNames(dds_subset_filt)

# timepoint2 (T+NK)_vs_timepoint1 (T)
deg_TpNK_tp2_vs_Tonly_tp1_results <- generateResults_upd(dds_object = dds_subset_filt, 
                                                             coeff_name = "condition_tp_Tumor_plus_NK_timepoint_2_vs_Tumor_only_timepoint_1",
                                                             cond_numerator = "Tumor_plus_NK_timepoint_2", 
                                                             cond_denominator = "Tumor_only_timepoint_1",
                                                             cond_variable="condition_tp")
CD_results <- deg_TpNK_tp2_vs_Tonly_tp1_results
openxlsx::write.xlsx(deg_TpNK_tp2_vs_Tonly_tp1_results, file = file.path(deg_dir, "deg_TpNK_tp2_vs_Tonly_tp1_CD_results.xlsx"))
```

```{r CD DE main - enrichment, message=FALSE, eval=TRUE, include=TRUE, cache=TRUE}
#msigdbr::msigdbr_species()
#packageVersion("msigdbr")
#msigdbr::msigdbr_collections()

# Loading MsigDB geneset collections ----
gs_hallmark <- hypeR::msigdb_gsets(species = "Mus musculus", category = c("H"), clean=TRUE) 
gs_C2_kegg <- hypeR::msigdb_gsets(species = "Mus musculus", category = c("C2"), subcategory = "CP:KEGG", clean=TRUE) 
gs_C5_GOBP <- hypeR::msigdb_gsets(species = "Mus musculus", category = c("C5"), subcategory = "GO:BP", clean=TRUE) 
gs_C5_GOCC <- hypeR::msigdb_gsets(species = "Mus musculus", category = c("C5"), subcategory = "GO:CC", clean=TRUE) 
gs_C5_GOMF <- hypeR::msigdb_gsets(species = "Mus musculus", category = c("C5"), subcategory = "GO:MF", clean=TRUE) 

TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOBP <- hypeR::hypeR(signature = deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$mgi_symbol, 
                                                      genesets = gs_C5_GOBP, 
                                                      test="hypergeometric", 
                                                      background=nrow(dds_subset_filt))
CD_GOBP <- TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOBP

TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOCC <- hypeR::hypeR(signature = deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$mgi_symbol, 
                                                      genesets = gs_C5_GOCC, 
                                                      test="hypergeometric", 
                                                      background=nrow(dds_subset_filt))
CD_GOCC <- TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOCC

TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOMF <- hypeR::hypeR(signature = deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$mgi_symbol, 
                                                      genesets = gs_C5_GOMF, 
                                                      test="hypergeometric", 
                                                      background=nrow(dds_subset_filt))
CD_GOMF <- TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOMF

#hypeR::hyp_show(hyp_obj$data$clA, simple = FALSE)
hypeR::hyp_to_excel(TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOBP, file_path=file.path(deg_dir, "TpNK_tp2_vs_Tonly_tp1_CD_enrichment_C5_GOBP.xlsx"))
hypeR::hyp_to_excel(TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOCC, file_path=file.path(deg_dir, "TpNK_tp2_vs_Tonly_tp1_CD_enrichment_C5_GOCC.xlsx"))
hypeR::hyp_to_excel(TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOMF, file_path=file.path(deg_dir, "TpNK_tp2_vs_Tonly_tp1_CD_enrichment_C5_GOMF.xlsx"))

# plotting enrichment results ----
TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOBP_plot <- hypeR::hyp_dots(TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOBP, merge=TRUE, fdr=0.05, top = 20, abrv=70, val="fdr", title="GOBP: T+NK timepoint2 vs T-only timepoint1 CD")+theme_bw()

TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOCC_plot <- hypeR::hyp_dots(TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOCC, merge=TRUE, fdr=0.05, top = 20, abrv=70, val="fdr", title="GOCC: T+NK timepoint2 vs T-only timepoint1 CD")+theme_bw()

TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOMF_plot <- hypeR::hyp_dots(TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOMF, merge=TRUE, fdr=0.05, top = 20, abrv=70, val="fdr", title="GOMF:T+NK timepoint2 vs T-only timepoint1 CD")+theme_bw()

ggsave(filename = paste0(deg_dir, "TpNK_tp2_vs_Tonly_tp1_AB_hyp_GOBP_plot.png"), 
       plot=TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOBP_plot,
       width = 20, height = 20, units = "cm")
ggsave(filename = paste0(deg_dir, "TpNK_tp2_vs_Tonly_tp1_AB_hyp_GOCC_plot.png"), 
       plot=TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOCC_plot,
       width = 20, height = 20, units = "cm")
ggsave(filename = paste0(deg_dir, "TpNK_tp2_vs_Tonly_tp1_AB_hyp_GOMF_plot.png"), 
       plot=TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOMF_plot,
       width = 20, height = 20, units = "cm")


###### For publication ######

ggsave(filename = paste0(deg_dir, "publication/", "TpNK_tp2_vs_Tonly_tp1_AB_hyp_GOBP_plot.png"), 
       plot=TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOBP_plot,
       width = 20, height = 20, units = "cm")

ggsave(filename = paste0(deg_dir, "publication/", "TpNK_tp2_vs_Tonly_tp1_AB_hyp_GOBP_plot.eps"), 
       plot=TpNK_tp2_vs_Tonly_tp1_enrichment_C5_GOBP_plot,
       width = 20, height = 20, units = "cm")
```

### Highlighting genes of interest

Volcano plot highlighting genes of interest:
```{r 03 CD 1 DE main - plotting results - volcano plot, message=FALSE, eval=TRUE, include=TRUE, cache=TRUE}
# volcano plot
# heatmap - corrected
genes_of_interest
genes_of_interest_exp9.4_9.7

TpNK_tp2_vs_Tonly_tp1_volcano_plot_1 <- plotVolcano(dds_results_obj = deg_TpNK_tp2_vs_Tonly_tp1_results$results_all, 
                                             genes_of_interest = names(genes_of_interest), 
                                             plot_title = "T+NK tp2 vs T-only tp1 CD genes set 1")

TpNK_tp2_vs_Tonly_tp1_volcano_plot_2 <- plotVolcano(dds_results_obj = deg_TpNK_tp2_vs_Tonly_tp1_results$results_all, 
                                             genes_of_interest = names(genes_of_interest_exp9.4_9.7), 
                                             plot_title = "T+NK tp2 vs T-only tp1 CD genes set 2")


ggsave(filename = paste0(deg_dir, "TpNK_tp2_vs_Tonly_tp1_CD_volcano_plot_gene_set_1.png"), 
       plot=TpNK_tp2_vs_Tonly_tp1_volcano_plot_1,
       width = 20, height = 20, units = "cm")

ggsave(filename = paste0(deg_dir, "TpNK_tp2_vs_Tonly_tp1_CD_volcano_plot_gene_set_2.png"), 
       plot=TpNK_tp2_vs_Tonly_tp1_volcano_plot_2,
       width = 20, height = 20, units = "cm")


##### for publication ######

TpNK_tp2_vs_Tonly_tp1_volcano_plot_2 <- plotVolcano_repel(dds_results_obj = deg_TpNK_tp2_vs_Tonly_tp1_results$results_all, 
                                             genes_of_interest = names(genes_of_interest_exp9.4_9.7), 
                                             plot_title = "T+NK tp2 vs T-only tp1 CD")

ggsave(filename = paste0(deg_dir,"publication/", "TpNK_tp2_vs_Tonly_tp1_CD_volcano_plot.png"), 
       plot=TpNK_tp2_vs_Tonly_tp1_volcano_plot_2,
       width = 18, height = 12, units = "cm")
ggsave(filename = paste0(deg_dir,"publication/", "TpNK_tp2_vs_Tonly_tp1_CD_volcano_plot.eps"), 
       plot=TpNK_tp2_vs_Tonly_tp1_volcano_plot_2,
       width = 18, height = 12, units = "cm")
```

```{r 03 CD 2 DE main - plotting results - volcano plot print, message=FALSE, eval=TRUE, include=TRUE, fig.dim = c(10, 10)}
print(TpNK_tp2_vs_Tonly_tp1_volcano_plot_1)
print(TpNK_tp2_vs_Tonly_tp1_volcano_plot_2)

```

Heatmap of normalized and scaled expression of `r length(deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$ensembl_id)` significantly differentially expressed genes before and after removeal of batch effects coming from different experiments:

```{r 03 CD DE main - plotting results - heatmap plot, message=FALSE, eval=TRUE, include=TRUE, cache=TRUE, fig.dim = c(12, 14)}
# heatmap and heatmap batch corrected - corrected 
metadata_heatmap <- as.data.frame(colData(dds_subset))

heatmap_counts <- SummarizedExperiment::assay(transf_object)
heatmap_counts_NObatch <- SummarizedExperiment::assay(transf_batch_NObatch_experiment) # removed xperiment batch effects! ? use experimen+cell_line removed???

heatmap_counts_NObatch_deg <- heatmap_counts_NObatch[rownames(heatmap_counts_NObatch) %in% deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$ensembl_id, ]

annotation_col <- metadata_heatmap %>%
  dplyr::select(experiment, cell_line_label, condition_tp) %>% # condition, timepoint_cell_harvesting, 
  dplyr::arrange(condition_tp, cell_line_label)

heatmap_counts_NObatch_deg_ord <- heatmap_counts_NObatch_deg[, match(rownames(annotation_col), colnames(heatmap_counts_NObatch_deg))]

ensembl2symbol_annot <- deg_TpNK_tp2_vs_Tonly_tp1_results$results_all %>%
  dplyr::select(ensembl_id, mgi_symbol)

color.scheme <- rev(RColorBrewer::brewer.pal(8,"RdBu")) # generate the color scheme to use

ann_colors = list(
  experiment = c(EXP9.4 = "#00262E", EXP9.5 = "#005f73", EXP9.6 = "#0a9396", EXP9.7 = "#94d2bd"),
  condition_tp = c(Tumor_only_timepoint_1 = "#DD3344", Tumor_plus_NK_timepoint_1 = "#FF9F1C", Tumor_plus_NK_timepoint_2 = "#553388"),
  cell_line_label = c(C = "#E9D8A6", D = "#D9BE6D")
)

heatmap_corrected <- pheatmap::pheatmap(heatmap_counts_NObatch_deg_ord,
                   main = "Heatmap of signif. DEG",
                   scale = "row",
                   annotation_col = annotation_col,
                   annotation_colors = ann_colors,
                   #annotation_row = row_annot_symbols,
                   show_colnames = FALSE,
                   show_rownames = FALSE,
                   cluster_cols = FALSE,
                   #cluster_rows = counts_deg_ord_row_cor_hclust,
                   color = color.scheme,
                   fontsize = 10, fontsize_row = 10 #height=10, cellwidth = 11, cellheight = 11
                   ) # 

ggsave(filename = paste0(deg_dir, "TpNK_tp2_vs_Tonly_tp1_CD_heatmap_batch_corrected.png"), 
       plot=heatmap_corrected,
       width = 20, height = 20, units = "cm")

###### For publication #####

ggsave(filename = paste0(deg_dir, "publication/", "TpNK_tp2_vs_Tonly_tp1_CD_heatmap_bathc_corrected.png"), 
       plot=heatmap_corrected,
       width = 20, height = 20, units = "cm")
ggsave(filename = paste0(deg_dir, "publication/", "TpNK_tp2_vs_Tonly_tp1_CD_heatmap_batch_corrected.eps"), 
       plot=heatmap_corrected,
       width = 20, height = 20, units = "cm")
```

```{r 03 CD "small heatmap"}

# identify the 25 gene set
minL2FC_Ids <- order(deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$log2FoldChange)[1:25]
maxL2FC_Ids <- order(decreasing = T, deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$log2FoldChange)[1:25]
L2FC_EMSEMBL <- deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$ensembl_id[c(minL2FC_Ids, maxL2FC_Ids)]
L2FC_EMSEMBL <- c(L2FC_EMSEMBL, "ENSMUSG00000075602", "ENSMUSG00000060675")

heatmap_counts_NObatch <- SummarizedExperiment::assay(transf_batch_NObatch_experiment) # removed xperiment batch effects! ? use experimen+cell_line removed???

heatmap_counts_NObatch_deg <- heatmap_counts_NObatch[rownames(heatmap_counts_NObatch) %in% L2FC_EMSEMBL, ]

annotation_col <- metadata_heatmap %>%
  dplyr::select(experiment, cell_line_label, condition_tp) %>% # condition, timepoint_cell_harvesting, 
  dplyr::arrange(condition_tp, cell_line_label)

heatmap_counts_NObatch_deg_ord <- heatmap_counts_NObatch_deg[, match(rownames(annotation_col), colnames(heatmap_counts_NObatch_deg))]

ensembl2symbol_annot <- deg_TpNK_tp2_vs_Tonly_tp1_results$results_all %>%
  dplyr::select(ensembl_id, mgi_symbol)

rownames(heatmap_counts_NObatch_deg_ord) <- ensembl2symbol_annot[match(x = row.names(heatmap_counts_NObatch_deg_ord), table = ensembl2symbol_annot$ensembl_id),]$mgi_symbol

color.scheme <- rev(RColorBrewer::brewer.pal(8,"RdBu")) # generate the color scheme to use

ann_colors = list(
  experiment = c(EXP9.4 = "#00262E", EXP9.5 = "#005f73", EXP9.6 = "#0a9396", EXP9.7 = "#94d2bd"),
  condition_tp = c(Tumor_only_timepoint_1 = "#DD3344", Tumor_plus_NK_timepoint_1 = "#FF9F1C", Tumor_plus_NK_timepoint_2 = "#553388"),
  cell_line_label = c(C = "#E9D8A6", D = "#D9BE6D")
)


heatmap_corrected <- pheatmap::pheatmap(heatmap_counts_NObatch_deg_ord,
                   main = "Heatmap of signif. DEG - Batch Corrected 50 most regulated genes",
                   scale = "row",
                   annotation_col = annotation_col,
                   annotation_colors = ann_colors,
                   #annotation_row = row_annot_symbols,
                   show_colnames = FALSE,
                   show_rownames = T,
                   cluster_cols = FALSE,
                   #cluster_rows = counts_deg_ord_row_cor_hclust,
                   color = color.scheme,
                   fontsize = 10, fontsize_row = 10 #height=10, cellwidth = 11, cellheight = 11
                   ) # 

##### For publication #####
ggsave(filename = paste0(deg_dir, "publication/", "TpNK_tp2_vs_Tonly_tp1_CD_heatmap_corrected_50_genes.eps"), 
       plot=heatmap_corrected,
       width = 24, height = 20, units = "cm")
ggsave(filename = paste0(deg_dir, "publication/", "TpNK_tp2_vs_Tonly_tp1_CD_heatmap_corrected_50_genes.png"), 
       plot=heatmap_corrected,
       width = 24, height = 20, units = "cm")
```

```{r 03 CD DE main - plotting results - heatmap plot - batchEffect experiment removed, message=FALSE, eval=TRUE, include=TRUE, cache=TRUE, fig.dim = c(12, 14)}
# heatmap and heatmap batch corrected - corrected 
metadata_heatmap <- as.data.frame(colData(dds_subset))

heatmap_counts <- SummarizedExperiment::assay(transf_object)
heatmap_counts_NObatch <- SummarizedExperiment::assay(transf_batch_NObatch_experiment) # removed xperiment batch effects! ? use experimen+cell_line removed???

heatmap_counts_deg <- heatmap_counts[rownames(heatmap_counts) %in% deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$ensembl_id, ]
heatmap_counts_NObatch_deg <- heatmap_counts_NObatch[rownames(heatmap_counts_NObatch) %in% deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$ensembl_id, ]

annotation_col <- metadata_heatmap %>%
  dplyr::select(experiment, cell_line_label, condition_tp) %>% # condition, timepoint_cell_harvesting, 
  dplyr::arrange(condition_tp, cell_line_label)

heatmap_counts_deg_ord <- heatmap_counts_deg[, match(rownames(annotation_col), colnames(heatmap_counts_deg))]
heatmap_counts_NObatch_deg_ord <- heatmap_counts_NObatch_deg[, match(rownames(annotation_col), colnames(heatmap_counts_NObatch_deg))]

ensembl2symbol_annot <- deg_TpNK_tp2_vs_Tonly_tp1_results$results_all %>%
  dplyr::select(ensembl_id, mgi_symbol)

color.scheme <- rev(RColorBrewer::brewer.pal(8,"RdBu")) # generate the color scheme to use

ann_colors = list(
  experiment = c(EXP9.4 = "#4682B4", EXP9.5 = "#7846B4", EXP9.6 = "#B47846", EXP9.7 = "#82B446", EXP9.8 = "grey")
)

heatmap_corrected <- pheatmap::pheatmap(heatmap_counts_NObatch_deg_ord,
                   main = "Heatmap of signif. DEG - removed experiment BatchEffect",
                   scale = "row",
                   annotation_col = annotation_col,
                   annotation_colors = ann_colors,
                   #annotation_row = row_annot_symbols,
                   show_colnames = FALSE,
                   show_rownames = FALSE,
                   cluster_cols = FALSE,
                   #cluster_rows = counts_deg_ord_row_cor_hclust,
                   color = color.scheme,
                   fontsize = 10, fontsize_row = 10 #height=10, cellwidth = 11, cellheight = 11
                   ) # 

ggsave(filename = paste0(deg_dir, "TpNK_tp2_vs_Tonly_tp1_CD_heatmap_corrected.png"), 
       plot=heatmap_corrected,
       width = 20, height = 20, units = "cm")
```

```{r  03 AB_CD "see the overlap of the two gene sets from AB and CD"}
AB_CD_UP_venn <- list("AB_up" = AB_results$results_signif[AB_results$results_signif$log2FoldChange > 0,]$mgi_symbol,
                   "CD_up" = CD_results$results_signif[CD_results$results_signif$log2FoldChange > 0,]$mgi_symbol)

AB_CD_DOWN_venn <- list("AB_down" = AB_results$results_signif[AB_results$results_signif$log2FoldChange < 0,]$mgi_symbol,
                   "CD_down" = CD_results$results_signif[CD_results$results_signif$log2FoldChange < 0,]$mgi_symbol)

library(ggvenn)
AB_CD_UP_venn_plot <- ggvenn(
  AB_CD_UP_venn, 
  fill_color = c("#E0E2E3", "#8EA7CE"),
  stroke_size = 0.5, set_name_size = 4
  )


AB_CD_DOWN_venn_plot <- ggvenn(
  AB_CD_DOWN_venn, 
  fill_color = c("#E0E2E3", "#8EA7CE"),
  stroke_size = 0.5, set_name_size = 4
  )


deg_name <- "TpNK_tp2_AB_CD_comparison"
deg_dir <- file.path(output_dir, paste0("deg_", deg_name, "/"))

intersect(AB_results$results_signif[AB_results$results_signif$log2FoldChange < 0,]$mgi_symbol,
          CD_results$results_signif[CD_results$results_signif$log2FoldChange < 0,]$mgi_symbol)

ggsave(filename = paste0(deg_dir, "AB_CD_UP_venn_plot.png"), 
       plot=AB_CD_UP_venn_plot,
       width = 20, height = 20, units = "cm")

ggsave(filename = paste0(deg_dir, "AB_CD_UP_venn_plot.eps"), 
       plot=AB_CD_UP_venn_plot,
       width = 20, height = 20, units = "cm")

ggsave(filename = paste0(deg_dir, "AB_CD_DOWN_venn_plot.png"), 
       plot=AB_CD_DOWN_venn_plot,
       width = 20, height = 20, units = "cm")

ggsave(filename = paste0(deg_dir, "AB_CD_DOWN_venn_plot.eps"), 
       plot=AB_CD_DOWN_venn_plot,
       width = 20, height = 20, units = "cm")


#aggreagated excel sheet for 
Deg_AB_xlsx <- openxlsx::read.xlsx("~/workspace/results_dir/nk_tum_immunoedit_complete/deg_TpNK_tp2_vs_Tonly_tp1_AB/deg_TpNK_tp2_vs_Tonly_tp1_AB_results.xlsx",sheet = 'results_signif')
Deg_CD_xlsx <- openxlsx::read.xlsx("~/workspace/results_dir/nk_tum_immunoedit_complete/deg_TpNK_tp2_vs_Tonly_tp1_CD/deg_TpNK_tp2_vs_Tonly_tp1_CD_results.xlsx",sheet = 'results_signif')


AB_specific <- Deg_AB_xlsx[Deg_AB_xlsx$mgi_symbol %in% setdiff(Deg_AB_xlsx$mgi_symbol, Deg_CD_xlsx$mgi_symbol),]%>%
  select("ensembl_id", "mgi_symbol", "mgi_description", "log2FoldChange", "padj")
CD_specific <- Deg_CD_xlsx[Deg_CD_xlsx$mgi_symbol %in% setdiff(Deg_CD_xlsx$mgi_symbol, Deg_AB_xlsx$mgi_symbol),]%>%
  select("ensembl_id", "mgi_symbol", "mgi_description", "log2FoldChange", "padj")

AB_intersect_tmp <-  Deg_AB_xlsx[which(Deg_AB_xlsx$mgi_symbol %in% intersect(Deg_AB_xlsx$mgi_symbol, Deg_CD_xlsx$mgi_symbol)), ] %>% 
  select("ensembl_id", "mgi_symbol", "mgi_description", "log2FoldChange", "padj") %>% 
  rename("AB_log2FoldChange" = "log2FoldChange", "AB_padj" = "padj")

CD_intersect_tmp <-  Deg_CD_xlsx[which(Deg_CD_xlsx$mgi_symbol %in% intersect(Deg_AB_xlsx$mgi_symbol, Deg_CD_xlsx$mgi_symbol)), ] %>% 
  select("ensembl_id", "mgi_symbol", "mgi_description", "log2FoldChange", "padj") %>% 
  rename("CD_log2FoldChange" = "log2FoldChange", "CD_padj" = "padj")

CD_intersect_tmp <- CD_intersect_tmp[match(AB_intersect_tmp$mgi_symbol, CD_intersect_tmp$mgi_symbol),] 

AB_CD_intersect <- cbind(AB_intersect_tmp, select(CD_intersect_tmp, "CD_log2FoldChange", "CD_padj"))

AB_CD_ABCD_diff_intersect <- list("AB_specific" = AB_specific,
                                  "CD_specific" = CD_specific,
                                  "AB_CD_intersect" = AB_CD_intersect)
openxlsx::write.xlsx(AB_CD_ABCD_diff_intersect, 
                     "~/workspace/results_dir/nk_tum_immunoedit_complete/deg_TpNK_tp2_AB_CD_comparison/AB_CD_diff_and_interactions.xlsx")



```

Make a small heatmap with genes, overlaping from AB and CD.

```{r  03  AB_CD "heatmap - overlap of genes from AB and CD"}

deg_name <- "TpNK_tp2_AB_CD_comparison"
deg_dir <- file.path(output_dir, paste0("deg_", deg_name, "/"))
dir.create(deg_dir)

precomputed_objects_deg_filename <- paste0(deg_name, "_dds_objects.RData")
precomputed_objects_deg_file <- file.path(deg_dir, precomputed_objects_deg_filename)

precomputed_transf_objects_deg_filename <- paste0(deg_name, "_log2_vsd_filt_objects.RData")
precomputed_transf_objects_deg_file <- file.path(deg_dir, precomputed_transf_objects_deg_filename)

experiment_subset <- c("EXP9.4", "EXP9.5", "EXP9.6", "EXP9.7")

AB_CD_intersect_genes <- intersect(AB_results$results_signif$mgi_symbol, CD_results$results_signif$mgi_symbol)

AB_res_subset <- AB_results$results_signif[AB_results$results_signif$mgi_symbol %in% AB_CD_intersect_genes,]
#CD_res_subset <- CD_results$results_signif[CD_results$results_signif$mgi_symbol %in% AB_CD_intersect_genes,]

AB_res_subset_UP <- AB_res_subset[AB_res_subset$log2FoldChange > 0,]
AB_res_subset_DOWN <- AB_res_subset[AB_res_subset$log2FoldChange < 0,]

#AB_res_subset_UP <- AB_res_subset_UP %>% arrange(desc(log2FoldChange)) %>% slice_head(n=20)
#AB_res_subset_DOWN <- AB_res_subset_DOWN %>% arrange(log2FoldChange) %>% slice_head(n=10)

#OR padj
AB_res_subset_UP <- AB_res_subset_UP %>% arrange(padj) %>% slice_head( n = 20)
AB_res_subset_DOWN <- AB_res_subset_DOWN %>% arrange(desc(padj)) %>% slice_head(n=10)

geneset <- c(AB_res_subset_UP$mgi_symbol, AB_res_subset_DOWN$mgi_symbol)


deg_name <- "TpNK_tp2_vs_Tonly_tp1_AB"
deg_dir <- file.path(output_dir, paste0("deg_", deg_name, "/"))

precomputed_objects_deg_filename <- paste0(deg_name, "_dds_objects.RData")
precomputed_objects_deg_file <- file.path(deg_dir, precomputed_objects_deg_filename)

precomputed_transf_objects_deg_filename <- paste0(deg_name, "_log2_vsd_filt_objects.RData")
precomputed_transf_objects_deg_file <- file.path(deg_dir, precomputed_transf_objects_deg_filename)

load(precomputed_objects_deg_file)
load(precomputed_transf_objects_deg_file)

transf_object <- vsd_subset_filt
transf_batch_NObatch_experiment <- vsd_subset_filt
transf_batch_NObatch_experiment_count <- limma::removeBatchEffect(SummarizedExperiment::assay(transf_batch_NObatch_experiment), transf_batch_NObatch_experiment$experiment)
SummarizedExperiment::assay(transf_batch_NObatch_experiment) <- transf_batch_NObatch_experiment_count

metadata_heatmap <- as.data.frame(colData(dds_subset))
heatmap_counts <- SummarizedExperiment::assay(transf_object)
heatmap_counts_NObatch <- SummarizedExperiment::assay(transf_batch_NObatch_experiment) # removed xperiment batch effects! ? use experimen+cell_line removed???

heatmap_counts_deg <- heatmap_counts[rownames(heatmap_counts) %in% deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$ensembl_id, ]
heatmap_counts_NObatch_deg <- heatmap_counts_NObatch[rownames(heatmap_counts_NObatch) %in% deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$ensembl_id, ]

annotation_col <- metadata_heatmap %>%
    dplyr::select(experiment, cell_line_label, condition_tp) %>% # condition, timepoint_cell_harvesting, 
    dplyr::arrange(condition_tp, cell_line_label)

heatmap_counts_deg_ord <- heatmap_counts_deg[, match(rownames(annotation_col), colnames(heatmap_counts_deg))]
heatmap_counts_NObatch_deg_ord <- heatmap_counts_NObatch_deg[, match(rownames(annotation_col), colnames(heatmap_counts_NObatch_deg))]


ensembl2symbol_annot <- deg_TpNK_tp2_vs_Tonly_tp1_results$results_all %>%
    dplyr::select(ensembl_id, mgi_symbol)

ensembl_gene_set <- ensembl2symbol_annot[which(ensembl2symbol_annot$mgi_symbol %in% geneset),]
ensembl_gene_set <- ensembl_gene_set[ match(geneset , ensembl_gene_set$mgi_symbol),]


indexes <- which(row.names(heatmap_counts_NObatch_deg_ord) %in% ensembl_gene_set$ensembl_id)
heatmap_counts_NObatch_deg_ord <- heatmap_counts_NObatch_deg_ord[indexes,]
heatmap_counts_NObatch_deg_ord <- heatmap_counts_NObatch_deg_ord[ match(ensembl_gene_set$ensembl_id , row.names(heatmap_counts_NObatch_deg_ord)),]


rownames(heatmap_counts_NObatch_deg_ord) <- ensembl_gene_set$mgi_symbol

color.scheme <- rev(RColorBrewer::brewer.pal(8,"RdBu")) # generate the color scheme to use

ann_colors = list(
    experiment = c( EXP9.5 = "#005f73", EXP9.6 = "#0a9396", EXP9.7 = "#94d2bd"),
    condition_tp = c(Tumor_only_timepoint_1 = "#DD3344", Tumor_plus_NK_timepoint_1 = "#FF9F1C",  Tumor_plus_NK_timepoint_2 = "#553388"),
    cell_line_label = c(A = "#E9D8A6", B = "#D9BE6D")
)

heatmap_corrected <- pheatmap::pheatmap(heatmap_counts_NObatch_deg_ord,
                                        main = "Heatmap of signif. DEG - Batch Corrected",
                                        scale = "row",
                                        annotation_col = annotation_col,
                                        annotation_colors = ann_colors,
                                        #annotation_row = row_annot_symbols,
                                        show_colnames = FALSE,
                                        show_rownames = TRUE,
                                        cluster_cols = FALSE,
                                        cluster_rows = FALSE,
                                        color = color.scheme,
                                        fontsize = 10, fontsize_row = 10 #height=10, cellwidth = 11, cellheight = 11
) 

ggsave(filename = paste0(deg_dir, "publication/", "TpNK_tp2_vs_Tonly_tp1_heatmap_corrected_small_shared_gene_set_padj.eps"), 
       plot=heatmap_corrected,
       width = 30, height = 14, units = "cm")

ggsave(filename = paste0(deg_dir, "publication/", "TpNK_tp2_vs_Tonly_tp1_heatmap_corrected_small_shared_gene_set_padj.png"), 
       plot=heatmap_corrected,
       width = 30, height = 14, units = "cm")



##### The same but for CD #####

deg_name <- "TpNK_tp2_vs_Tonly_tp1_CD"
deg_dir <- file.path(output_dir, paste0("deg_", deg_name, "/"))

precomputed_objects_deg_filename <- paste0(deg_name, "_dds_objects.RData")
precomputed_objects_deg_file <- file.path(deg_dir, precomputed_objects_deg_filename)

precomputed_transf_objects_deg_filename <- paste0(deg_name, "_log2_vsd_filt_objects.RData")
precomputed_transf_objects_deg_file <- file.path(deg_dir, precomputed_transf_objects_deg_filename)

load(precomputed_objects_deg_file)
load(precomputed_transf_objects_deg_file)

transf_object <- vsd_subset_filt
transf_batch_NObatch_experiment <- vsd_subset_filt
transf_batch_NObatch_experiment_count <- limma::removeBatchEffect(SummarizedExperiment::assay(transf_batch_NObatch_experiment), transf_batch_NObatch_experiment$experiment)
SummarizedExperiment::assay(transf_batch_NObatch_experiment) <- transf_batch_NObatch_experiment_count

metadata_heatmap <- as.data.frame(colData(dds_subset))
heatmap_counts <- SummarizedExperiment::assay(transf_object)
heatmap_counts_NObatch <- SummarizedExperiment::assay(transf_batch_NObatch_experiment) # removed xperiment batch effects! ? use experimen+cell_line removed???

heatmap_counts_deg <- heatmap_counts[rownames(heatmap_counts) %in% deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$ensembl_id, ]
heatmap_counts_NObatch_deg <- heatmap_counts_NObatch[rownames(heatmap_counts_NObatch) %in% deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$ensembl_id, ]

annotation_col <- metadata_heatmap %>%
    dplyr::select(experiment, cell_line_label, condition_tp) %>% # condition, timepoint_cell_harvesting, 
    dplyr::arrange(condition_tp, cell_line_label)

heatmap_counts_deg_ord <- heatmap_counts_deg[, match(rownames(annotation_col), colnames(heatmap_counts_deg))]
heatmap_counts_NObatch_deg_ord <- heatmap_counts_NObatch_deg[, match(rownames(annotation_col), colnames(heatmap_counts_NObatch_deg))]


ensembl2symbol_annot <- deg_TpNK_tp2_vs_Tonly_tp1_results$results_all %>%
    dplyr::select(ensembl_id, mgi_symbol)

ensembl_gene_set <- ensembl2symbol_annot[which(ensembl2symbol_annot$mgi_symbol %in% geneset),]
ensembl_gene_set <- ensembl_gene_set[ match(geneset , ensembl_gene_set$mgi_symbol),]


indexes <- which(row.names(heatmap_counts_NObatch_deg_ord) %in% ensembl_gene_set$ensembl_id)
heatmap_counts_NObatch_deg_ord <- heatmap_counts_NObatch_deg_ord[indexes,]
heatmap_counts_NObatch_deg_ord <- heatmap_counts_NObatch_deg_ord[ match(ensembl_gene_set$ensembl_id , row.names(heatmap_counts_NObatch_deg_ord)),]


rownames(heatmap_counts_NObatch_deg_ord) <- ensembl_gene_set$mgi_symbol

color.scheme <- rev(RColorBrewer::brewer.pal(8,"RdBu")) # generate the color scheme to use

ann_colors = list(
  experiment = c(EXP9.4 = "#00262E", EXP9.5 = "#005f73", EXP9.6 = "#0a9396", EXP9.7 = "#94d2bd"),
  condition_tp = c(Tumor_only_timepoint_1 = "#DD3344", Tumor_plus_NK_timepoint_1 = "#FF9F1C", Tumor_plus_NK_timepoint_2 = "#553388"),
  cell_line_label = c(C = "#E9D8A6", D = "#D9BE6D")
)


heatmap_corrected <- pheatmap::pheatmap(heatmap_counts_NObatch_deg_ord,
                                        main = "Heatmap of signif. DEG - Batch Corrected",
                                        scale = "row",
                                        annotation_col = annotation_col,
                                        annotation_colors = ann_colors,
                                        #annotation_row = row_annot_symbols,
                                        show_colnames = FALSE,
                                        show_rownames = TRUE,
                                        cluster_cols = FALSE,
                                        cluster_rows = FALSE,
                                        color = color.scheme,
                                        
                                        fontsize = 10, fontsize_row = 10 #height=10, cellwidth = 11, cellheight = 11
) 

ggsave(filename = paste0(deg_dir, "publication/", "TpNK_tp2_vs_Tonly_tp1_heatmap_corrected_small_shared_gene_set_padj.eps"), 
       plot=heatmap_corrected,
       width = 30, height = 14, units = "cm")

ggsave(filename = paste0(deg_dir, "publication/", "TpNK_tp2_vs_Tonly_tp1_heatmap_corrected_small_shared_gene_set_padj.png"), 
       plot=heatmap_corrected,
       width = 30, height = 14, units = "cm")

```


```{r  03 "try to plot in the big, integrated DATASET"}

#First, I need to load full Dataset:

experiment_name="nk_tum_immunoedit_complete"
output_dir <- "/home/rstudio/workspace/results_dir/nk_tum_immunoedit_complete/" 
precomputed_objects_filename <- paste0(experiment_name, "_dds_objects.RData")
precomputed_objects_file <- file.path(output_dir, precomputed_objects_filename)

# log2 and vsd transformed
precomputed_transf_objects_filename <- paste0(experiment_name, "_log2_vsd_filt_objects.RData")
precomputed_transf_objects_file <- file.path(output_dir, precomputed_transf_objects_filename)

# check if object loaded if not load
if(!exists(precomputed_objects_file)){
    load(precomputed_objects_file)
    load(precomputed_transf_objects_file)
} 
experiment_subset <- c("EXP9.4", "EXP9.5", "EXP9.6", "EXP9.7")
condition_tp_subset <- c("Tumor_only_timepoint_1",
                         "Tumor_plus_NK_timepoint_2",
                         "Tumor_plus_NK_timepoint_1",
                         "Tumor_plus_WT_NK_timepoint_2")
abs_filt_samples=3
deg_design <- as.formula("~ experiment + cell_line_label + condition_tp")


dds_subset <- dds_filt[ , dds_filt$condition_tp %in% condition_tp_subset]  
dds_subset <- dds_subset[ , dds_subset$experiment %in% experiment_subset]
dds_subset$condition_tp <- droplevels(dds_subset$condition_tp)
table(dds_subset$condition_tp)

dds_subset$experiment <- droplevels(dds_subset$experiment)
table(dds_subset$experiment)

dds_subset$cell_line_label <- droplevels(dds_subset$cell_line_label)
table(dds_subset$cell_line_label)
dds_subset_filt <- filterDatasets(dds_subset, 
                                  abs_filt = TRUE, 
                                  abs_filt_samples = abs_filt_samples) # at least in N samples, which is a smallest group size

cat("... dds\n")
design(dds_subset_filt) <- deg_design
dds_subset_filt <- DESeq2::estimateSizeFactors(dds_subset_filt)
dds_subset_filt <- DESeq2::DESeq(dds_subset_filt) # do not replace outliers based on replicates
log2_norm_subset_filt <- DESeq2::normTransform(dds_subset_filt)
vsd_subset_filt <- DESeq2::vst(dds_subset_filt, blind = FALSE) 
transf_batch_NObatch_experiment <- vsd_subset_filt
transf_batch_NObatch_experiment_count <- limma::removeBatchEffect(SummarizedExperiment::assay(transf_batch_NObatch_experiment), transf_batch_NObatch_experiment$experiment)
SummarizedExperiment::assay(transf_batch_NObatch_experiment) <- transf_batch_NObatch_experiment_count
deg_TpNK_tp2_vs_Tonly_tp1_results <- generateResults_upd(dds_object = dds_subset_filt, 
                                                         coeff_name = "condition_tp_Tumor_plus_NK_timepoint_2_vs_Tumor_only_timepoint_1",
                                                         cond_numerator = "Tumor_plus_NK_timepoint_2", 
                                                         cond_denominator = "Tumor_only_timepoint_1",
                                                         cond_variable="condition_tp")
 deg_TpNK_tp2_vs_Tonly_tp1_results

metadata_heatmap <- as.data.frame(colData(dds_subset))
transf_object <- vsd_subset_filt
heatmap_counts <- SummarizedExperiment::assay(transf_object)
heatmap_counts_NObatch <- SummarizedExperiment::assay(transf_batch_NObatch_experiment) # removed xperiment batch effects! ? use experimen+cell_line removed???

heatmap_counts_NObatch_deg <- heatmap_counts_NObatch[rownames(heatmap_counts_NObatch) %in% deg_TpNK_tp2_vs_Tonly_tp1_results$results_signif$ensembl_id, ]

annotation_col <- metadata_heatmap %>%
  dplyr::select(experiment, cell_line_label, condition_tp) %>% # condition, timepoint_cell_harvesting, 
  dplyr::arrange(condition_tp, cell_line_label)

heatmap_counts_NObatch_deg_ord <- heatmap_counts_NObatch_deg[, match(rownames(annotation_col), colnames(heatmap_counts_NObatch_deg))]

ensembl2symbol_annot <- deg_TpNK_tp2_vs_Tonly_tp1_results$results_all %>%
  dplyr::select(ensembl_id, mgi_symbol)


# select top 20 and 10 gens from ab cd up and down regulated
AB_res_path <- "~/workspace/results_dir/nk_tum_immunoedit_complete/deg_TpNK_tp2_vs_Tonly_tp1_AB/deg_TpNK_tp2_vs_Tonly_tp1_AB_results.xlsx"
CD_res_path <- "~/workspace/results_dir/nk_tum_immunoedit_complete/deg_TpNK_tp2_vs_Tonly_tp1_CD/deg_TpNK_tp2_vs_Tonly_tp1_CD_results.xlsx"

AB_results <- openxlsx::read.xlsx(xlsxFile = AB_res_path, sheet = "results_signif")
CD_results <- openxlsx::read.xlsx(xlsxFile = CD_res_path, sheet = "results_signif")

AB_CD_intersect_genes <- intersect(AB_results$mgi_symbol, CD_results$mgi_symbol)

AB_res_subset <- AB_results[AB_results$mgi_symbol %in% AB_CD_intersect_genes,]
CD_res_subset <- CD_results[CD_results$mgi_symbol %in% AB_CD_intersect_genes,]

combined_AB_CD <- rbind.data.frame(AB_res_subset[,1:14], CD_res_subset[,1:14])

AB_CD_res_subset_UP <- combined_AB_CD[combined_AB_CD$log2FoldChange > 0,]
AB_CD_res_subset_DOWN <- combined_AB_CD[combined_AB_CD$log2FoldChange < 0,]

AB_CD_res_subset_UP <- AB_CD_res_subset_UP %>% arrange(padj)
AB_CD_res_subset_UP <- AB_CD_res_subset_UP[!duplicated(AB_CD_res_subset_UP$mgi_symbol),] %>% slice_head(n=20)

AB_CD_res_subset_DOWN <- AB_CD_res_subset_DOWN %>% arrange(padj)
AB_CD_res_subset_DOWN <- AB_CD_res_subset_DOWN[!duplicated(AB_CD_res_subset_DOWN$mgi_symbol),]%>% slice_head(n=10)

geneset <- c(AB_CD_res_subset_UP$mgi_symbol, AB_CD_res_subset_DOWN$mgi_symbol)


ensembl_gene_set <- ensembl2symbol_annot[which(ensembl2symbol_annot$mgi_symbol %in% geneset),]
ensembl_gene_set <- ensembl_gene_set[ match(geneset , ensembl_gene_set$mgi_symbol),]

indexes <- which(row.names(heatmap_counts_NObatch_deg_ord) %in% ensembl_gene_set$ensembl_id)
heatmap_counts_NObatch_deg_ord <- heatmap_counts_NObatch_deg_ord[indexes,]
heatmap_counts_NObatch_deg_ord <- heatmap_counts_NObatch_deg_ord[ match(ensembl_gene_set$ensembl_id , row.names(heatmap_counts_NObatch_deg_ord)),]


rownames(heatmap_counts_NObatch_deg_ord) <- ensembl_gene_set$mgi_symbol


######

color.scheme <- rev(RColorBrewer::brewer.pal(8,"RdBu")) # generate the color scheme to use

ann_colors = list(
  experiment = c(EXP9.4 = "#00262E", EXP9.5 = "#005f73", EXP9.6 = "#0a9396", EXP9.7 = "#94d2bd"),
  condition_tp = c(Tumor_only_timepoint_1 = "#DD3344", Tumor_plus_NK_timepoint_1 = "#FF9F1C", Tumor_plus_NK_timepoint_2 = "#553388"),
  cell_line_label = c( A = "#E9D8A6", B = "#D9BE6D", C = "#676F54", D = "#395B50")
)

heatmap_corrected <- pheatmap::pheatmap(heatmap_counts_NObatch_deg_ord,
                   main = "Heatmap of signif. DEG",
                   scale = "row",
                   annotation_col = annotation_col,
                   annotation_colors = ann_colors,
                   #annotation_row = row_annot_symbols,
                   show_colnames = FALSE,
                   cluster_cols = FALSE,
                   cluster_rows = FALSE,
                   show_rownames = TRUE,
                   color = color.scheme,
                   fontsize = 10, fontsize_row = 10 #height=10, cellwidth = 11, cellheight = 11
                   ) # 

ggsave(filename = paste0("~/workspace/results_dir/nk_tum_immunoedit_complete/deg_TpNK_tp2_AB_CD_comparison/", "TpNK_tp2_vs_Tonly_tp1_heatmap_corrected_small_shared_gene_set_padj.eps"), 
       plot=heatmap_corrected,
       width = 30, height = 14, units = "cm")

ggsave(filename = paste0("~/workspace/results_dir/nk_tum_immunoedit_complete/deg_TpNK_tp2_AB_CD_comparison/", "TpNK_tp2_vs_Tonly_tp1_heatmap_corrected_small_shared_gene_set_padj.png"), 
       plot=heatmap_corrected,
       width = 30, height = 14, units = "cm")

```